---
title: "PopPsiSeq Dev1"
author: "Charlie Soeder"
date: "11/14/2018"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(root.dir = '/Users/csoeder/Research/PSIseq/PopPsiSeq/')
knitr::opts_knit$set(root.dir='/Users/csoeder/Research/PSIseq/PopPsiSeq/')
library("tidyverse")
library("knitr")


```

## 16 Nov 2018
```{r include=FALSE}
library("yaml")
getwd()
trammel <- read_yaml("config.yaml")
data_sets.df <- plyr::ldply(trammel$data_sets, data.frame)
data_sets.df$name <- as.factor(data_sets.df$name)
data_sets.df$paired<- as.factor(data_sets.df$paired)
data_sets.df$experimental<- as.factor(data_sets.df$experimental)
data_sets.df$species<- as.factor(data_sets.df$species)
data_sets.df$source<- as.factor(data_sets.df$source)
```

experimental data:

```{r echo=FALSE}
data_sets.df %>%  filter( !is.na(experimental)) %>%  select(c(name, paired, experimental, source)) %>%  arrange(experimental, desc(name)) %>% kable(caption="Sequenced Experimental Samples")
```

Population-wide sample count by species:

```{r echo=FALSE, results='asis'}
data_sets.df %>%  filter( !is.na(species)) %>% group_by(species) %>% summarise(sample_count=n()) %>% kable(caption="Number of Sequenced Samples per Species")
```


load & discuss FASTP summary


```{r echo=FALSE}

fastp_summary <- read_delim("meta/sequenced_reads.dat", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(fastp_summary ) <- c("name","type","measure","value")
fastp_summary$name <- as.factor(fastp_summary$name)
fastp_summary$type <- as.factor(fastp_summary$type)
fastp_summary$measure <- as.factor(fastp_summary$measure)


#fastp_summary %>%  filter(type=="prefiltered" | type=="postfiltered") %>% filter(measure == "total_reads")


```

prefilt:

```{r echo=FALSE}

prefiltered_stats <- inner_join(fastp_summary %>%  filter(type=="prefiltered"), data_sets.df, by=c("name"="name"))

prefiltered_stats %>% filter(measure=='total_reads') %>%  summarise(minimum = min(value), average=mean(value) , maximum = max(value)) #filter(measure=="total_reads") %>% 


filtration_stats <- inner_join(fastp_summary %>%  filter(type=="prefiltered" | type == 'postfiltered'), data_sets.df, by=c("name"="name"))
filtration_stats$type <- factor(filtration_stats$type, levels=c("prefiltered", "postfiltered"))

pre_post_counts <- filtration_stats %>% filter(measure=='total_reads') %>%  group_by(type)  %>%  summarise(minimum = min(value), average=mean(value) , maximum = max(value)) 
retention_percent <- filtration_stats %>% filter(measure=='total_reads') %>% select(c(name,type,value)) %>%  spread(type,value) %>% mutate(retention=100*postfiltered/prefiltered) %>%  summarise(type='percent retention', minimum = min(retention), average=mean(retention) , maximum = max(retention)) 
rbind(pre_post_counts, retention_percent) %>% kable()
```

```{r echo=FALSE}
ggplot(filtration_stats %>% filter(measure == "total_reads")) + geom_line(aes(group=name, x=type,y=value)) +  geom_point(aes(x=type, y = value)) 
```

```{r echo=FALSE}

ggplot(filtration_stats %>% filter(measure == "q30_rate")) + geom_line(aes(group=name, x=type,y=value)) +  geom_point(aes(x=type, y = value)) 

```


```{r echo=FALSE}

dupe_stats <- inner_join(fastp_summary %>% filter(type=='duplication' & measure =='rate') %>%  mutate(percent=value*100) %>% select(c(name,percent)), data_sets.df, by=c("name"="name"))

```


## 19 Nov 2018

load and discuss bam summary

depth of coverage is effed???

```{r echo=FALSE}
vs_droSim1.bwa <- read_delim("meta/alignments.vs_droSim1.bwa.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(vs_droSim1.bwa) <- c("sample","measure","value")
vs_droSim1.bwa$aligner <- as.factor("bwa")
vs_droSim1.bwa$sample <- as.factor(vs_droSim1.bwa$sample)
vs_droSim1.bwa$measure <- as.factor(vs_droSim1.bwa$measure)


vs_droSim1.bwaUniq <- read_delim("meta/alignments.vs_droSim1.bwaUniq.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(vs_droSim1.bwaUniq) <- c("sample","measure","value")
vs_droSim1.bwaUniq$aligner <- as.factor("bwaUniq")
vs_droSim1.bwaUniq$sample <- as.factor(vs_droSim1.bwaUniq$sample)
vs_droSim1.bwaUniq$measure <- as.factor(vs_droSim1.bwaUniq$measure)
```

```{r echo=FALSE}

vs_droSim1.bwa %>%  filter(measure=='total_read_count' | measure == 'total_mapped_count')
vs_droSim1.bwaUniq %>%  filter( measure == 'total_mapped_count') %>% mutate(measure='filtered_mapped_count')




before_After.counts <- rbind(vs_droSim1.bwa %>%  filter(measure=='total_read_count' | measure == 'total_mapped_count'), vs_droSim1.bwaUniq %>%  filter( measure == 'total_mapped_count') %>% mutate(measure='filtered_mapped_count'))

before_After.counts$measure <- factor(before_After.counts$measure, levels = c('total_read_count','total_mapped_count','filtered_mapped_count'))

ggplot(before_After.counts) + geom_line(aes(group=sample, x=measure,y=value)) + geom_point(aes(x=measure, y=value, group=sample))


```

```{r echo=FALSE}
before_After.counts.aug <- before_After.counts %>% select(-c(aligner)) %>% spread(measure, value) %>%  mutate(mapping_retention=total_mapped_count/total_read_count, filter_retention = filtered_mapped_count/total_mapped_count)

before_After.counts.aug %>% gather(total_read_count:filtered_mapped_count, key="measure", value="value") %>%  group_by(measure) %>% summarise(minimum = min(value), average=mean(value), median = median(value), maximum = max(value))


```





```{r echo=FALSE}
before_After.cov <- inner_join( vs_droSim1.bwa %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), vs_droSim1.bwaUniq %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), by='sample', suffix=c(".before",".after") ) %>%  mutate(depth_retention = avg_depth.after/avg_depth.before, breadth_retention=total_breadth.after/total_breadth.before)
```

```{r echo=FALSE}

ggplot(before_After.counts) + geom_line(aes(group=sample, x=measure,y=value)) + geom_line(aes(group=sample, x=measure,y=value)) + geom_point(aes(x=measure, y=value, group=sample,color=sample)) + labs(title="Read Counts by Processing Step: Raw, Mapped, Filtered", x="", y="Number Reads" ) + scale_y_log10()

```


## 20 Nov 2018

Depth of coverage:

```{r echo=FALSE}

#before_After.cov %>% gather(avg_depth.before:breadth_retention, key=measure, value=value)

covstats.dpth <- before_After.cov %>% gather(avg_depth.before, key=measure, value=value) %>% summarise(step="pre-filtration depth",minimum = min(value), average=mean(value), median = median(value), maximum = max(value))

covstats.dpth <- rbind(covstats.dpth, before_After.cov %>% gather(avg_depth.after, key=measure, value=value)  %>% summarise(step="post-filtration depth",minimum = min(value), average=mean(value), median = median(value), maximum = max(value)))


covstats.dpth <- rbind(covstats.dpth, before_After.cov %>% gather(depth_retention, key=measure, value=value)  %>% summarise(step="depth retention",minimum = min(value), average=mean(value), median = median(value), maximum = max(value)))

covstats.dpth %>% kable(caption="Average Depth of Coverage for Raw and Filtered Alignments")


##report by subdivision:
before_After.cov.meta <- inner_join(before_After.cov, data_sets.df, by=c('sample'='name'))
before_After.cov.meta %>% group_by(species)  %>% gather(avg_depth.before, key=measure, value=value) %>% summarise(step="pre-filtration depth",minimum = min(value), average=mean(value), median = median(value), maximum = max(value)) 


```

```{r echo=FALSE}

before_After.cov.gathered.meta <- inner_join(before_After.cov %>%  gather(avg_depth.before:breadth_retention, key="measure", value="value") , data_sets.df, by=c("sample"="name") ) 

ggplot(before_After.cov.gathered.meta %>% filter(measure== "avg_depth.before" | measure== "avg_depth.after") %>% mutate(measure=factor(measure, levels=c("avg_depth.before" , "avg_depth.after" )))) + geom_line(aes(group=sample, x=measure,y=value, color=experimental)) + geom_point(aes(x=measure, y=value, group=sample)) + labs(title="Depth Of Coverage for Raw and Filtered Alignments", x="", y="Reads Per BP, Genome-Wide" )


```



```{r echo=FALSE}

covstats.brth <- before_After.cov %>% gather(total_breadth.before, key=measure, value=value) %>% summarise(step="pre-filtration breadth",minimum = 100*min(value), average=100*mean(value), median = 100*median(value), maximum = 100*max(value))

covstats.brth <- rbind(covstats.brth, before_After.cov %>% gather(total_breadth.after, key=measure, value=value)  %>% summarise(step="post-filtration breadth",minimum = 100*min(value), average=100*mean(value), median = 100*median(value), maximum = 100*max(value)))


covstats.brth <- rbind(covstats.brth, before_After.cov %>% gather(breadth_retention, key=measure, value=value)  %>% summarise(step="breadth retention",minimum = 100*min(value), average=100*mean(value), median = 100*median(value), maximum = 100*max(value)))

covstats.brth %>% kable(caption="Breadth of Coverage Statistics for Raw and Filtered Alignments", digits=1)

```




```{r echo=FALSE}


ggplot(before_After.cov.gathered.meta %>% filter(measure== "total_breadth.before" | measure== "total_breadth.after") %>% mutate(measure=factor(measure, levels=c("total_breadth.before" , "total_breadth.after" )))) + geom_line(aes(group=sample, x=measure,y=value, color=experimental)) + geom_point(aes(x=measure, y=value, group=sample)) + labs(title="Breadth Of Coverage for Raw and Filtered Alignments", x="", y="Percentage of Genome Mapped To" )
```


## 27 Nov 2018

better kable-tables with prettyNum() and sitools::f2si

https://stackoverflow.com/questions/3245862/format-numbers-to-significant-figures-nicely-in-r

sitools:
https://stackoverflow.com/questions/11340444/is-there-an-r-function-to-format-number-using-unit-prefix


```{r echo=FALSE}
library("sitools")

human_readable_croncher <- function(num_in) {
	dig <- 3
	num_out <- formatC(num_in, digits=dig, format='g') %>% as.numeric() %>% f2si()
	return(num_out)
}

```

```{r echo=FALSE}

rbind(pre_post_counts, retention_percent) %>%  mutate( minimum=human_readable_croncher(minimum), average=human_readable_croncher(average), maximum=human_readable_croncher(maximum)) %>% kable(caption="Read Counts by Sample")

```


```{r echo=FALSE}
covstats.brth  %>% mutate( minimum=human_readable_croncher(minimum), average=human_readable_croncher(average), maximum=human_readable_croncher(maximum)) %>% kable(caption="Breadth of Coverage Statistics for Raw and Filtered Alignments")
```

Also, need to add panels by reference genome. 

Also, some mention of reference genomes in the summary, with stats?


First, clean up the summarizers with a loading wrapper function
```{r echo=FALSE, include=FALSE}

bam_summary_loader <- function(filename, aligner="bwa", reference='dm6'){
	
	tmp.df <- read_delim(filename, "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
	names(tmp.df) <- c("sample","measure","value")
	
	tmp.df$sample <- as.factor(tmp.df$sample)
	tmp.df$measure <- as.factor(tmp.df$measure)
	tmp.df$aligner <- as.factor(aligner)
	tmp.df$reference <- as.factor(reference)
	
	return(tmp.df)
	
}

vs_droSim1.bwa <- bam_summary_loader(filename = "meta/alignments.vs_droSim1.bwa.summary",aligner="bwa", reference="droSim1")
vs_droSec1.bwa <- bam_summary_loader(filename = "meta/alignments.vs_droSec1.bwa.summary",aligner="bwa", reference="droSec1")
vs_droSim1.bwaUniq <- bam_summary_loader(filename = "meta/alignments.vs_droSim1.bwaUniq.summary",aligner="bwaUniq", reference="droSim1")
vs_droSec1.bwaUniq <- bam_summary_loader(filename = "meta/alignments.vs_droSec1.bwaUniq.summary",aligner="bwaUniq", reference="droSec1")

all_alignments <- rbind(vs_droSim1.bwa,vs_droSec1.bwa,vs_droSim1.bwaUniq,vs_droSec1.bwaUniq)


```


Previous stuff still works:

```{r echo=FALSE}

before_After.counts <- rbind(vs_droSim1.bwa %>%  filter(measure=='total_read_count' | measure == 'total_mapped_count'), vs_droSim1.bwaUniq %>%  filter( measure == 'total_mapped_count') %>% mutate(measure='filtered_mapped_count'))

before_After.counts$measure <- factor(before_After.counts$measure, levels = c('total_read_count','total_mapped_count','filtered_mapped_count'))

ggplot(before_After.counts) + geom_line(aes(group=sample, x=measure,y=value)) + geom_line(aes(group=sample, x=measure,y=value)) + geom_point(aes(x=measure, y=value, group=sample,color=sample)) + labs(title="Read Counts by Processing Step: Raw, Mapped, Filtered", x="", y="Number Reads" ) + scale_y_log10()

```

```{r echo=FALSE}
ggplot(all_alignments %>%  filter(measure=="avg_depth")) +geom_line(aes(x=aligner,y=value, group=sample), color="black")+ geom_point(aes(x=aligner,y=value, group=sample, color=sample)) + facet_grid(.~reference) + labs(title="Depth of Coverage")
```

```{r echo=FALSE}
ggplot(all_alignments %>%  filter(measure=="total_breadth"))+geom_line(aes(x=aligner,y=value, group=sample), color="black") + geom_point(aes(x=aligner,y=value, group=sample, color=sample)) + facet_grid(.~reference)+ labs(title="Breadth of Coverage")
```

## 28 Nov 2018

Retooling some diagrams and pipes



```{r echo=FALSE}

bam_sum_test <- all_alignments %>%  filter( (measure=='total_read_count' & aligner=="bwa") | measure == 'total_mapped_count' ) %>% mutate(measure=ifelse(aligner=="bwaUniq", "filtered_mapped_count", ifelse(measure=="total_read_count","total_read_count","total_mapped_count"))) 
bam_sum_test$measure <- factor(bam_sum_test$measure, levels = c('total_read_count','total_mapped_count','filtered_mapped_count'))

ggplot(bam_sum_test) + geom_line(aes(group=sample, x=measure,y=value)) + geom_point(aes(x=measure, y=value, group=sample,color=sample))+ facet_grid(.~reference) + labs(title="Read Counts by Processing Step: Raw, Mapped, Filtered", x="", y="Number Reads" ) + scale_y_log10()



bam_sum_test.spread <- bam_sum_test %>% select(-c(aligner)) %>%  spread(measure, value) %>%  mutate(mapping_retention=total_mapped_count/total_read_count, filter_retention = filtered_mapped_count/total_mapped_count)

bam_sum_test.spread %>% gather(total_read_count:filtered_mapped_count, key="measure", value="value") %>%  group_by(measure) %>% summarise(minimum = min(value), average=mean(value), median = median(value), maximum = max(value)) %>% mutate(minimum = human_readable_croncher(minimum), average=human_readable_croncher(average), median = human_readable_croncher(median), maximum = human_readable_croncher(maximum)) %>% kable(caption="Read Counts During Alignment & Filtration")
```

We can easily break down the table further with a second grouping:

```{r echo=FALSE}

bam_sum_test.spread %>% gather(total_read_count:filtered_mapped_count, key="measure", value="value") %>%  group_by(measure, reference) %>% summarise(minimum = min(value), average=mean(value), median = median(value), maximum = max(value)) %>% mutate(minimum = human_readable_croncher(minimum), average=human_readable_croncher(average), median = human_readable_croncher(median), maximum = human_readable_croncher(maximum)) %>% kable(caption="Read Counts During Alignment & Filtration")
```

using spread and gather to clean up this mess:
```{r echo=FALSE, include=FALSE}

before_After.cov <- inner_join( vs_droSim1.bwa %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), vs_droSim1.bwaUniq %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), by='sample', suffix=c(".before",".after") ) %>%  mutate(depth_retention = avg_depth.after/avg_depth.before, breadth_retention=total_breadth.after/total_breadth.before)
```

```{r echo=FALSE, include=FALSE}

depth.process <- all_alignments %>%  filter(measure=='avg_depth' ) %>% spread(aligner, value) %>%  mutate(depth_retention = 100*bwaUniq/bwa) %>% rename(before=bwa, after=bwaUniq)

covstats.dpth <- depth.process %>% summarise(step="pre-filtration depth",minimum = min(before), average=mean(before), median = median(before), maximum = max(before))

covstats.dpth <- rbind(covstats.dpth, depth.process %>% summarise(step="post-filtration depth",minimum = min(after), average=mean(after), median = median(after), maximum = max(after)))


covstats.dpth <- rbind(covstats.dpth, depth.process  %>% summarise(step="depth retention percent",minimum = min(depth_retention), average=mean(depth_retention), median = median(depth_retention), maximum = max(depth_retention)))


covstats.dpth %>% kable(caption="Depth of Coverage Statistics for Raw and Filtered Alignments", digits=1 )

```

```{r echo=FALSE, include=FALSE}
ggplot(depth.process %>%  select(-c(depth_retention)) %>%  gather(before:after, key=measure, value=value) %>% mutate(measure=factor(measure, levels=c("before","after"))) ) + geom_point(aes(group=sample, x=measure,y=value, color=sample)) + geom_line(aes(group=sample, x=measure,y=value)) + facet_grid(.~reference)

```

Again, just group_by() for a more detailed breakdown:

```{r echo=FALSE, include=FALSE}
depth.process %>% group_by(reference)%>% summarise(step="pre-filtration depth",minimum = min(before), average=mean(before), median = median(before), maximum = max(before)) 
```


## 29 Nov 2018



```{r echo=FALSE, include=FALSE}

breadth.process <- all_alignments %>%  filter(measure=='total_breadth' ) %>% spread(aligner, value) %>%  mutate(breadth_retention = 100*bwaUniq/bwa) %>% rename(before=bwa, after=bwaUniq)


covstats.brdth <- breadth.process %>% summarise(step="pre-filtration breadth",minimum = 100*min(before), average=100*mean(before), median = 100*median(before), maximum = 100*max(before))

covstats.brdth <- rbind(covstats.brdth, breadth.process %>% summarise(step="post-filtration breadth",minimum = 100*min(after), average=100*mean(after), median = 100*median(after), maximum = 100*max(after)))

covstats.brdth <- rbind(covstats.brdth, breadth.process  %>% summarise(step="breadth retention percent",minimum = min(breadth_retention), average=mean(breadth_retention), median = median(breadth_retention), maximum = max(breadth_retention)))

covstats.brdth %>% kable(caption="Breadth of Coverage Statistics for Raw and Filtered Alignments", digits=1 )

```


```{r echo=FALSE, include=FALSE}
ggplot(breadth.process %>%  select(-c(breadth_retention)) %>%  gather(before:after, key=measure, value=value) %>% mutate(measure=factor(measure, levels=c("before","after"))) ) + geom_point(aes(group=sample, x=measure,y=value, color=sample)) + geom_line(aes(group=sample, x=measure,y=value)) + facet_grid(.~reference)

```

do this to include script contents eg in the methods
```{r  engine='bash', comment='', tidy=TRUE, tidy.opts=list(width.cutoff=60)} 
cat scripts/bam_summarizer.py
```

yikes, looks like i might need to run a pep8 check LOL

VCFs are done building:

 cat all_samples.vs_droSim1.bwaUniq.vcf | head -n 1000 > all_samples.vs_droSim1.bwaUniq.vcf.subset










