---
title: "PopPsiSeq Dev1"
author: "Charlie Soeder"
date: "11/14/2018"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(root.dir = '/Users/csoeder/Research/PSIseq/PopPsiSeq/')
knitr::opts_knit$set(root.dir='/Users/csoeder/Research/PSIseq/PopPsiSeq/')
library("tidyverse")
library("knitr")


```

## 16 Nov 2018
```{r include=FALSE}
library("yaml")
getwd()
trammel <- read_yaml("config.yaml")
data_sets.df <- plyr::ldply(trammel$data_sets, data.frame)
data_sets.df$name <- as.factor(data_sets.df$name)
data_sets.df$paired<- as.factor(data_sets.df$paired)
data_sets.df$experimental<- as.factor(data_sets.df$experimental)
data_sets.df$species<- as.factor(data_sets.df$species)
data_sets.df$source<- as.factor(data_sets.df$source)
```

experimental data:

```{r echo=FALSE}
data_sets.df %>%  filter( !is.na(experimental)) %>%  select(c(name, paired, experimental, source)) %>%  arrange(experimental, desc(name)) %>% kable(caption="Sequenced Experimental Samples")
```

Population-wide sample count by species:

```{r echo=FALSE, results='asis'}
data_sets.df %>%  filter( !is.na(species)) %>% group_by(species) %>% summarise(sample_count=n()) %>% kable(caption="Number of Sequenced Samples per Species")
```


load & discuss FASTP summary


```{r echo=FALSE}

fastp_summary <- read_delim("meta/sequenced_reads.dat", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(fastp_summary ) <- c("name","type","measure","value")
fastp_summary$name <- as.factor(fastp_summary$name)
fastp_summary$type <- as.factor(fastp_summary$type)
fastp_summary$measure <- as.factor(fastp_summary$measure)


#fastp_summary %>%  filter(type=="prefiltered" | type=="postfiltered") %>% filter(measure == "total_reads")


```

prefilt:

```{r echo=FALSE}

prefiltered_stats <- inner_join(fastp_summary %>%  filter(type=="prefiltered"), data_sets.df, by=c("name"="name"))

prefiltered_stats %>% filter(measure=='total_reads') %>%  summarise(minimum = min(value), average=mean(value) , maximum = max(value)) #filter(measure=="total_reads") %>% 


filtration_stats <- inner_join(fastp_summary %>%  filter(type=="prefiltered" | type == 'postfiltered'), data_sets.df, by=c("name"="name"))
filtration_stats$type <- factor(filtration_stats$type, levels=c("prefiltered", "postfiltered"))

pre_post_counts <- filtration_stats %>% filter(measure=='total_reads') %>%  group_by(type)  %>%  summarise(minimum = min(value), average=mean(value) , maximum = max(value)) 
retention_percent <- filtration_stats %>% filter(measure=='total_reads') %>% select(c(name,type,value)) %>%  spread(type,value) %>% mutate(retention=100*postfiltered/prefiltered) %>%  summarise(type='percent retention', minimum = min(retention), average=mean(retention) , maximum = max(retention)) 
rbind(pre_post_counts, retention_percent) %>% kable()
```

```{r echo=FALSE}
ggplot(filtration_stats %>% filter(measure == "total_reads")) + geom_line(aes(group=name, x=type,y=value)) +  geom_point(aes(x=type, y = value)) 
```

```{r echo=FALSE}

ggplot(filtration_stats %>% filter(measure == "q30_rate")) + geom_line(aes(group=name, x=type,y=value)) +  geom_point(aes(x=type, y = value)) 

```


```{r echo=FALSE}

dupe_stats <- inner_join(fastp_summary %>% filter(type=='duplication' & measure =='rate') %>%  mutate(percent=value*100) %>% select(c(name,percent)), data_sets.df, by=c("name"="name"))

```


## 19 Nov 2018

load and discuss bam summary

depth of coverage is effed???

```{r echo=FALSE}
vs_droSim1.bwa <- read_delim("meta/alignments.vs_droSim1.bwa.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(vs_droSim1.bwa) <- c("sample","measure","value")
vs_droSim1.bwa$aligner <- as.factor("bwa")
vs_droSim1.bwa$sample <- as.factor(vs_droSim1.bwa$sample)
vs_droSim1.bwa$measure <- as.factor(vs_droSim1.bwa$measure)


vs_droSim1.bwaUniq <- read_delim("meta/alignments.vs_droSim1.bwaUniq.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(vs_droSim1.bwaUniq) <- c("sample","measure","value")
vs_droSim1.bwaUniq$aligner <- as.factor("bwaUniq")
vs_droSim1.bwaUniq$sample <- as.factor(vs_droSim1.bwaUniq$sample)
vs_droSim1.bwaUniq$measure <- as.factor(vs_droSim1.bwaUniq$measure)
```

```{r echo=FALSE}

vs_droSim1.bwa %>%  filter(measure=='total_read_count' | measure == 'total_mapped_count')
vs_droSim1.bwaUniq %>%  filter( measure == 'total_mapped_count') %>% mutate(measure='filtered_mapped_count')




before_After.counts <- rbind(vs_droSim1.bwa %>%  filter(measure=='total_read_count' | measure == 'total_mapped_count'), vs_droSim1.bwaUniq %>%  filter( measure == 'total_mapped_count') %>% mutate(measure='filtered_mapped_count'))

before_After.counts$measure <- factor(before_After.counts$measure, levels = c('total_read_count','total_mapped_count','filtered_mapped_count'))

ggplot(before_After.counts) + geom_line(aes(group=sample, x=measure,y=value)) + geom_point(aes(x=measure, y=value, group=sample))


```

```{r echo=FALSE}
before_After.counts.aug <- before_After.counts %>% select(-c(aligner)) %>% spread(measure, value) %>%  mutate(mapping_retention=total_mapped_count/total_read_count, filter_retention = filtered_mapped_count/total_mapped_count)

before_After.counts.aug %>% gather(total_read_count:filtered_mapped_count, key="measure", value="value") %>%  group_by(measure) %>% summarise(minimum = min(value), average=mean(value), median = median(value), maximum = max(value))


```





```{r echo=FALSE}
before_After.cov <- inner_join( vs_droSim1.bwa %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), vs_droSim1.bwaUniq %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), by='sample', suffix=c(".before",".after") ) %>%  mutate(depth_retention = avg_depth.after/avg_depth.before, breadth_retention=total_breadth.after/total_breadth.before)
```

```{r echo=FALSE}

ggplot(before_After.counts) + geom_line(aes(group=sample, x=measure,y=value)) + geom_line(aes(group=sample, x=measure,y=value)) + geom_point(aes(x=measure, y=value, group=sample,color=sample)) + labs(title="Read Counts by Processing Step: Raw, Mapped, Filtered", x="", y="Number Reads" ) + scale_y_log10()

```


## 20 Nov 2018

Depth of coverage:

```{r echo=FALSE}

#before_After.cov %>% gather(avg_depth.before:breadth_retention, key=measure, value=value)

covstats.dpth <- before_After.cov %>% gather(avg_depth.before, key=measure, value=value) %>% summarise(step="pre-filtration depth",minimum = min(value), average=mean(value), median = median(value), maximum = max(value))

covstats.dpth <- rbind(covstats.dpth, before_After.cov %>% gather(avg_depth.after, key=measure, value=value)  %>% summarise(step="post-filtration depth",minimum = min(value), average=mean(value), median = median(value), maximum = max(value)))


covstats.dpth <- rbind(covstats.dpth, before_After.cov %>% gather(depth_retention, key=measure, value=value)  %>% summarise(step="depth retention",minimum = min(value), average=mean(value), median = median(value), maximum = max(value)))

covstats.dpth %>% kable(caption="Average Depth of Coverage for Raw and Filtered Alignments")


##report by subdivision:
before_After.cov.meta <- inner_join(before_After.cov, data_sets.df, by=c('sample'='name'))
before_After.cov.meta %>% group_by(species)  %>% gather(avg_depth.before, key=measure, value=value) %>% summarise(step="pre-filtration depth",minimum = min(value), average=mean(value), median = median(value), maximum = max(value)) 


```

```{r echo=FALSE}

before_After.cov.gathered.meta <- inner_join(before_After.cov %>%  gather(avg_depth.before:breadth_retention, key="measure", value="value") , data_sets.df, by=c("sample"="name") ) 

ggplot(before_After.cov.gathered.meta %>% filter(measure== "avg_depth.before" | measure== "avg_depth.after") %>% mutate(measure=factor(measure, levels=c("avg_depth.before" , "avg_depth.after" )))) + geom_line(aes(group=sample, x=measure,y=value, color=experimental)) + geom_point(aes(x=measure, y=value, group=sample)) + labs(title="Depth Of Coverage for Raw and Filtered Alignments", x="", y="Reads Per BP, Genome-Wide" )


```



```{r echo=FALSE}

covstats.brth <- before_After.cov %>% gather(total_breadth.before, key=measure, value=value) %>% summarise(step="pre-filtration breadth",minimum = 100*min(value), average=100*mean(value), median = 100*median(value), maximum = 100*max(value))

covstats.brth <- rbind(covstats.brth, before_After.cov %>% gather(total_breadth.after, key=measure, value=value)  %>% summarise(step="post-filtration breadth",minimum = 100*min(value), average=100*mean(value), median = 100*median(value), maximum = 100*max(value)))


covstats.brth <- rbind(covstats.brth, before_After.cov %>% gather(breadth_retention, key=measure, value=value)  %>% summarise(step="breadth retention",minimum = 100*min(value), average=100*mean(value), median = 100*median(value), maximum = 100*max(value)))

covstats.brth %>% kable(caption="Breadth of Coverage Statistics for Raw and Filtered Alignments", digits=1)

```




```{r echo=FALSE}


ggplot(before_After.cov.gathered.meta %>% filter(measure== "total_breadth.before" | measure== "total_breadth.after") %>% mutate(measure=factor(measure, levels=c("total_breadth.before" , "total_breadth.after" )))) + geom_line(aes(group=sample, x=measure,y=value, color=experimental)) + geom_point(aes(x=measure, y=value, group=sample)) + labs(title="Breadth Of Coverage for Raw and Filtered Alignments", x="", y="Percentage of Genome Mapped To" )
```






