---
title: "Simulans-Mauritiana Backcross Analysis for Amanda Moehring/Tom Hsiang"
author: "Charlie Soeder"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: true
    toc_depth: 5
    number_sections: true
  html_document: default
bibliography: moehring_references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_chunk$set(root.dir = '/Users/csoeder/Research/PSIseq/PopPsiSeq/')
#knitr::opts_knit$set(root.dir='/Users/csoeder/Research/PSIseq/PopPsiSeq/')
knitr::opts_knit$set(root.dir='/proj/cdjones_lab/csoeder/new_psiseq_project/')
#knitr::opts_knit$set(root.dir=peaDubDee)
library("tidyverse")
#require(devtools)
#install_version("ggplot2", version = "3.1.0", repos = "http://cran.us.r-project.org")
library("knitr")
library("gt")
library("yaml")
library("ggbio")
library("readr")
library("reshape2")
library("ggdendro")
library("dendextend")
library("grid")
library("gridExtra")
library("ComplexHeatmap")
library("rtracklayer")

#library("sitools")

# gotta set the working directory.....
```

```{r include=FALSE}

human_readable_croncher <- function(num_in) {
	dig <- 3
	num_out <- formatC(num_in, digits=dig, format='g') %>% as.numeric() %>% sitools::f2si()
	return(num_out)
}

bam_summary_loader <- function(filename, aligner="bwa", reference='dm6'){
	
	tmp.df <- read_delim(filename, "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
	names(tmp.df) <- c("sample","measure","value")
	
	tmp.df$sample <- as.factor(tmp.df$sample)
	tmp.df$measure <- as.factor(tmp.df$measure)
	tmp.df$aligner <- as.factor(aligner)
	tmp.df$reference <- as.factor(reference)
	
	return(tmp.df)
	
}


freqShift_loader <- function(filename, refgen = "dm6", p1 = "sim", p2= "sec"){
  
  windowedFreqShift.bg <- import.bedGraph(filename)#, genome=refgen)

  enhancement_col <-  paste("sum_",p1,"_deltaF", sep = "")
  depletion_col <- paste("sum_",p2,"_deltaF", sep="")

  names(mcols(windowedFreqShift.bg)) <-  c("score", enhancement_col, depletion_col, "num_snp")


  mcols(windowedFreqShift.bg)[[enhancement_col]] <- as.numeric(mcols(windowedFreqShift.bg)[[enhancement_col]])
  mcols(windowedFreqShift.bg)[[depletion_col]] <- as.numeric(mcols(windowedFreqShift.bg)[[depletion_col]])

  mcols(windowedFreqShift.bg)[[paste("avg_",p1,"_deltaF", sep = "")]] <- mcols(windowedFreqShift.bg)[[enhancement_col]]/windowedFreqShift.bg$num_snp 

  mcols(windowedFreqShift.bg)[[paste("avg_",p2,"_deltaF", sep = "")]] <- mcols(windowedFreqShift.bg)[[depletion_col]]/windowedFreqShift.bg$num_snp 
  windowedFreqShift.bg$win <- seq(1,nrow(mcols(windowedFreqShift.bg)))

  return(windowedFreqShift.bg)
}

psiSeq_loader <- function(filename){
  
  windowedPsiSeq.bg <- import.bedGraph(filename)#, genome=refgen)

  names( mcols( windowedPsiSeq.bg)) <- c("score", "shared","total")
mcols( windowedPsiSeq.bg)$shared <- as.numeric(mcols(windowedPsiSeq.bg)$shared)
mcols( windowedPsiSeq.bg)$total <- as.numeric(mcols(windowedPsiSeq.bg)$total)
windowedPsiSeq.bg$ratio <-windowedPsiSeq.bg$shared / windowedPsiSeq.bg$total

  return(windowedPsiSeq.bg)
}


#Earley2013_10A.PsiSeq2.shared_with_dSecFrag.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/dm6/bwaUniq/10A.SNPs_shared_with.fragSimulated_dSec1.vs_droSim1.bwaUniq.lifted_to_dm6.genomeWindowed_w100000_s100000.bed")

fig_cnt <- 0 
tbl_cnt <- 0 

```

# Introduction

> We essentially did introgression of loci linked to interspecies hybrid sterility between sim and mau for ten generations in both directions of backcross, then sequenced. We then looked for which regions of introgressed genome were associated with 10th generation sterile male offspring, compared to fertile male offspring. The data analysis was done ok, but missing some key elements that we really should add, and I also think there's a lot more precision that could be retrieved out of that data set. (Amanda Moehring, email to Corbin 17 July 2023)

> The project aims to map genes associated with interspecies hybrid male sterility between D. simulans and D. mauritiana. We selected using a sterile sperm morphology (we call "needle-eye". NE), selected across 10 generations of backcross by using sisters of sterile males in the backcross crosses. In the 10th generation, approx 50% of males were sterile and 50% fertile, which led us to think it could be a single locus. So we pooled the males of each phenotype into single samples for sequencing. The samples are as follows:

```         
mauGFP - wildtype D. mauritania with a GFP-tagged protamine (makes sperm fluoresce green).
simGFP - same as above, but D. simulans
BCM10NE - 10th generation backcross mauritiana males with needle-eye (sterile) sperm
BCM10WT - 10th generation backcross mauritiana males with wildtype sperm
BCS10NE - 10th generation backcross simulans males with needle-eye (sterile) sperm
BCS10WT - 10th generation backcross simulans males with wildtype sperm
```

-Amanda Moehring, email to me 15 Aug 2023

> Each pooled sample consisted of 30 individuals.

-Amanda Moehring, email to me 10 Oct 2023

# Materials, Methods, Data, Software

```{r include=FALSE}

trammel <- read_yaml("config.yaml")

```

## Reference Genomes

```{r include=FALSE, echo=FALSE}
ref_genomes.cfg.df <- plyr::ldply(trammel$reference_genomes, data.frame)

reference_genomes_summary.df  <- read_delim("data/summaries/reference_genomes/reference_genomes.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)

names(reference_genomes_summary.df) <- c("refGenome","measure","value")

reference_genomes_summary.df <- inner_join(ref_genomes.cfg.df,reference_genomes_summary.df,by=c("name"="refGenome")) 

```

The droSim1 reference genome was downloaded in FASTA format from UCSC Genome Browser; the UC Irvine mauritiana assembly (GCF_004382145.1) and the Princeton simulans assembly (GCF_016746395.2) were downloaded from NCBI. The droSim1 reference was the best-consolidated. Assemblies for the specific strains being investigated were provided by Tom Hsiang.

```{r echo=FALSE}
tbl_cnt <- tbl_cnt +1

reference_genomes_summary.df %>% spread(key=measure, value = value) %>% group_by(species) %>% select(-c("path","fai")) %>% gt()  %>% tab_header(title=paste("Table ",tbl_cnt, ". Size and Consolidation of Reference Genomes", sep = ""), subtitle=" ") %>%   fmt_number(columns = c(number_contigs, number_bases), suffixing = TRUE, decimals=0) %>% cols_label(name = " ", number_bases = "# bases", number_contigs= "# contigs")


```

The main chromosomes correspond to the following contigs in the NCBI reference:

```         
2L      NC_052520.2
2R      NC_052521.2
3L      NC_052522.2
3R      NC_052523.2
4       NC_052524.2
X       NC_052525.2
```

<https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_016746395.2/>

## Reference Annotations

NCBI

```{r echo=FALSE, message=FALSE, warning=FALSE}

annotations.df <- plyr::ldply(trammel$annotations, data.frame)
annotations.df$name <- as.factor(annotations.df$name)
annotations.df$genome <- as.factor(annotations.df$genome)

```

```{r echo=FALSE, message=FALSE, warning=FALSE }

ref_ann.stats <- read_delim("data/summaries/reference_annotations/reference_annotations.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(ref_ann.stats) <- c("annot", "measure", "type", "value")

tbl_cnt <- tbl_cnt + 1
ref_ann.stats.gt <- inner_join(ref_ann.stats  %>%   filter(type == 'total' | type == 'avg') %>% unite("measure", c("type", "measure"), sep = " ") %>% spread(measure, value),annotations.df  , by=c("annot"="name")) %>% gt() %>% tab_header(title=paste("Table ",tbl_cnt, ". Reference Annotations and their Sizes", sep = ""), subtitle= md("&nbsp;") ) %>% fmt_number(columns = c("avg size", "total count", "total size"), decimals = 1, drop_trailing_zeros = T, suffixing = T ) %>% tab_spanner(label = "size (bp)", columns = c("avg size", "total size")) %>% cols_label(`avg size` = "average", `total size` = "total") %>% cols_hide(c("gtf_path", "link","instructions"))

ref_ann.stats.gt

#write(ref_ann.stats.gt %>%  as_raw_html(), paste("results/tables/tbl",tbl_cnt,"_refAnnotationSummary.html", sep=""))

```

## Sequenced Reads

```{r include=FALSE, echo = FALSE}
data_sets.df <- plyr::ldply(trammel$data_sets, data.frame)
data_sets.df$name <- as.factor(data_sets.df$name)
data_sets.df$paired<- as.factor(data_sets.df$paired)
data_sets.df$source<- as.factor(data_sets.df$source)
data_sets.df$genealogy<- as.factor(data_sets.df$genealogy)
data_sets.df$genotype<- as.factor(data_sets.df$genotype)
```

A backcross and introgression experiment was performed, in which simulans-mauritiana crosses were backcrossed to each of their parent species, with and without selection for a mutant phenotype causing male sterility. The offspring of 10-generations of backcrossing were sequenced, as well as the parent stock.

```{r echo=FALSE}
tbl_cnt <- tbl_cnt + 1 
data_sets.df %>% filter(subgroups=='all') %>%  select(c(name, genealogy, genotype))  %>% gt()  %>% tab_header(title=paste("Table ",tbl_cnt, ".  Control and Selection Experiments", sep = ""), subtitle=" ") 



```

PsiSeq 1 and 2 use head-to-head comparisons of individual samples; PopPsiSeq compares subgroups (possibly consisting of a single individual) of samples.

```{r echo=FALSE}


contrast_samples.df <- data_sets.df %>% mutate(sample=name) %>% dplyr::select(sample, subgroups) %>% mutate(present=1)  %>% unique() %>% spread(key=subgroups, value=present, fill=0 ) 

contrast_samples.mat <- contrast_samples.df %>% dplyr::select(-c(sample)) %>% as.matrix()
rownames(contrast_samples.mat) <- contrast_samples.df$sample

colors = structure(c("white","black"), names = c(0, 1))

column_ha = columnAnnotation(foo = anno_text(colnames(t(contrast_samples.mat)), gp = gpar(fontsize = 6), rot = 90))


contrast_samples.hm <- Heatmap(t(contrast_samples.mat),clustering_method_rows = "single",clustering_method_columns = "single",show_column_dend = FALSE,show_row_dend = FALSE, col=colors, show_heatmap_legend = FALSE,bottom_annotation = column_ha, show_column_names = FALSE,rect_gp = gpar(col = "grey", lwd = 2), row_title = "subgroup", column_title="sample")


fig_cnt <- fig_cnt + 1

draw(contrast_samples.hm, column_title=paste("Figure ",fig_cnt,". Subgroup Definitions for PopPsiSeq", sep="")) 



```

### Pre-processing

```{r echo=FALSE, include=FALSE}
fastp_summary <- read_delim("data/summaries/intermediate/FASTP/all.sequenced_reads.dat", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(fastp_summary ) <- c("name","type","measure","value")
fastp_summary$name <- as.factor(fastp_summary$name)
fastp_summary$type <- as.factor(fastp_summary$type)
fastp_summary$measure <- as.factor(fastp_summary$measure)
```

```{r echo=FALSE, include=FALSE}
filtration_stats <- inner_join(fastp_summary %>%  filter(type=="prefiltered" | type == 'postfiltered'), data_sets.df, by=c("name"="name"))
filtration_stats$type <- factor(filtration_stats$type, levels=c("prefiltered", "postfiltered"))
```

These reads were preprocessed with FASTP [@Chen2018] for quality control and analytics.

Starting FASTQ files contained a total of $`r sum( filtration_stats %>% filter(type =='prefiltered') %>% filter(measure=='total_reads') %>% select(value) ) %>% human_readable_croncher() `$ reads; after QC, this dropped to $`r sum( filtration_stats %>% filter(type =='postfiltered') %>% filter(measure=='total_reads') %>% select(value) ) %>% human_readable_croncher() `$.

```{r echo=FALSE}
pre_post_counts <- filtration_stats %>% filter(measure=='total_reads') %>%  group_by(type)  %>%  summarise(minimum = min(value), average=mean(value) , maximum = max(value)) 
retention_percent <- filtration_stats %>% filter(measure=='total_reads') %>%  filter(subgroups=="all")%>% select(c(name,type,value)) %>%  spread(type,value) %>% mutate(retention=100*postfiltered/prefiltered) %>%  summarise(type='percent retention', minimum = min(retention), average=mean(retention) , maximum = max(retention))
```

```{r echo=FALSE}

tbl_cnt <- tbl_cnt+ 1 
rbind(pre_post_counts, retention_percent)  %>% gt()  %>% tab_header(title=paste("Table ",tbl_cnt, ". Read Retention Rate during Preprocessing", sep = ""), subtitle= md("&nbsp;")) %>%  fmt_number(columns = c(minimum, average,maximum),suffixing = TRUE, decimals=0) %>% cols_label(type=" ")
```

Filtration also increased the read quality, as seen in the increase in the fraction of reads with an average quality score \> 30:

```{r echo=FALSE}
fig_cnt <-fig_cnt +  1 
ggplot(filtration_stats %>% filter(measure == "q30_rate")) + geom_line(aes(group=name, x=type,y=100*value)) +  geom_point(aes(x=type, y = 100*value, color=name)) + labs(title = paste( "Figure ", fig_cnt, ". Percent of Reads with a mean QUAL > 30", sep = ""), y="Percent QUAL > 30", x="") + theme_clear()

```

Duplicate reads were also detected; these will be filtered during alignment:

```{r echo=FALSE, include=FALSE}
dupe_stats <- inner_join(fastp_summary %>% filter(type=='duplication' & measure =='rate') %>%  mutate(percent=value*100) %>% select(c(name,percent)), data_sets.df, by=c("name"="name"))
```

```{r echo=FALSE}
tbl_cnt <- tbl_cnt + 1

duplicationStats.gt <- dupe_stats %>%  summarise(minimum = min(percent), average=mean(percent), median=median(percent) , maximum = max(percent)) %>% gt() %>% tab_header(title=paste("Table ",tbl_cnt, ". Percentage Duplication", sep = ""), subtitle="FASTP estimate") %>% fmt_number(columns=c(minimum,median,average,maximum), decimals=1, )

duplicationStats.gt

#write(duplicationStats.gt %>%  as_raw_html(), paste("results/tables/tbl",tbl_cnt,"_duplicationStats.html", sep=""))


```

```{r echo=FALSE}
fig_cnt <- fig_cnt + 1 
ggplot(dupe_stats) + geom_histogram(aes(x=percent), bins=15) + labs(title=paste("Figure ", fig_cnt, ". Duplication Histogram", sep = ""), x="Read Duplication Rate (FASTP estimate)", y="Number Samples") + theme_clear()
```

## Mapped Reads

Reads were first mapped to a reference genome using the BWA SAMPE/SE algorithm. Then, the alignment file was filtered for uniqueness (ie, a read must be aligned optimally with no alternative or runner-up hits, `XT:A:U.*X0:i:1.*X1:i:0`), mapping/sequencing quality (`-q 20 -F 0x0100 -F 0x0200 -F 0x0300 -F 0x04`), and deduplication. This filtered alignment is called "bwaUniq".

```{r echo=FALSE, include=FALSE}

vs_droSim1.bwa <- bam_summary_loader(filename = "data/summaries/intermediate/BAMs/all.vs_droSim1.bwa.summary",aligner="bwa", reference="droSim1")
vs_ncbiMau.bwa <- bam_summary_loader(filename = "data/summaries/intermediate/BAMs/all.vs_ncbiMau.bwa.summary",aligner="bwa", reference="ncbiMau")
vs_droSim1.bwaUniq <- bam_summary_loader(filename = "data/summaries/intermediate/BAMs/all.vs_droSim1.bwaUniq.summary",aligner="bwaUniq", reference="droSim1")

vs_ncbiMau.bwaUniq <- bam_summary_loader(filename = "data/summaries/intermediate/BAMs/all.vs_ncbiMau.bwaUniq.summary",aligner="bwaUniq", reference="ncbiMau")

all_alignments <- rbind(vs_droSim1.bwa,vs_ncbiMau.bwa,vs_droSim1.bwaUniq,vs_ncbiMau.bwaUniq)
```

### Read & Alignment Quality

```{r echo=FALSE, warning = FALSE}

fig_cnt <- fig_cnt + 1 

readcount_process <- all_alignments %>%  filter( (measure=='total_read_count' & aligner=="bwa") | measure == 'total_mapped_count' ) %>% mutate(measure=ifelse(aligner=="bwaUniq", "filtered_mapped_count", ifelse(measure=="total_read_count","total_read_count","total_mapped_count"))) 
readcount_process$measure <- factor(readcount_process$measure, levels = c('total_read_count','total_mapped_count','filtered_mapped_count'))

ggplot(readcount_process) + geom_line(aes(group=sample, x=measure,y=value)) + geom_point(aes(x=measure, y=value, group=sample,color=sample)) + facet_grid(reference ~ . ) + labs(title=paste("Figure ", fig_cnt, ". Read Counts by Processing Step: Unmapped, Mapped, Filtered", sep = ""), x="", y="Number Reads" ) + scale_y_log10()  + theme_clear() + theme(axis.text.x = element_text(angle = -45, hjust = 0)) 




```

```{r echo=FALSE, warning=FALSE}
readcount_process.spread <- readcount_process %>% select(-c(aligner)) %>%  spread(measure, value) %>%  mutate(mapping_retention=total_mapped_count/total_read_count, filter_retention = filtered_mapped_count/total_mapped_count)

tbl_cnt <- tbl_cnt + 1 
readcount_process.spread%>% gather(total_read_count:filtered_mapped_count, key="measure", value="value") %>%  group_by(measure, reference) %>% summarise(minimum = min(value), average=mean(value), median = median(value), maximum = max(value)) %>%  gt() %>% tab_header(title=paste("Table ",tbl_cnt, ". Read Counts During Alignment & Filtration", sep = ""), subtitle=" ") %>% fmt_number(columns=c(minimum,median,average,maximum), decimals=1, suffixing = T)#  %>% mutate(minimum = human_readable_croncher(minimum), average=human_readable_croncher(average), median = human_readable_croncher(median), maximum = human_readable_croncher(maximum)) %>% kable(caption="")

```

The fraction of reads retained at each point:

```{r echo=FALSE, warning=FALSE}

tbl_cnt <- tbl_cnt + 1
readcount_process.spread %>% gather(mapping_retention:filter_retention, key="measure", value="value") %>%  group_by(measure, reference) %>% summarise(minimum = min(value), average=mean(value), median = median(value), maximum = max(value)) %>% gt()  %>%  tab_header(title=paste("Table ",tbl_cnt, ". Percentage of Reads Retained at Each Step", sep = ""), subtitle=" ") %>% fmt_percent(columns=c(minimum,median,average,maximum), decimals=1)#  %>% mutate(minimum = human_readable_croncher(minimum), average=human_readable_croncher(average), median = human_readable_croncher(median), maximum = human_readable_croncher(maximum)) %>% kable(caption="")
 #%>% mutate(minimum = human_readable_croncher(minimum), average=human_readable_croncher(average), median = human_readable_croncher(median), maximum = human_readable_croncher(maximum)) %>%  kable(caption="",digits=1)

```

### Depth & Breadth of Coverage

```{r echo=FALSE, include=FALSE}

before_After.cov <- inner_join( vs_droSim1.bwa %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), vs_droSim1.bwaUniq %>%  filter(measure=='avg_depth' | measure == 'total_breadth') %>% select(-c(aligner)) %>% spread(measure, value), by='sample', suffix=c(".before",".after") ) %>%  mutate(depth_retention = avg_depth.after/avg_depth.before, breadth_retention=total_breadth.after/total_breadth.before)


before_After.cov.gathered.meta <- inner_join(before_After.cov %>%  gather(avg_depth.before:breadth_retention, key="measure", value="value") , data_sets.df, by=c("sample"="name") ) 


```

Depth of coverage, ie, the genome-wide average number of mapped reads per base pair:

```{r echo=FALSE}

depth.process <- all_alignments  %>%  filter(measure=='avg_depth' )%>% spread(aligner, value) %>%  mutate(depth_retention = bwaUniq/bwa) %>% rename( bwa = "before", bwaUniq = "after")

covstats.dpth <- depth.process %>% group_by(reference)%>% summarise(step="pre-filtration depth",minimum = min(before), average=mean(before), median = median(before), maximum = max(before))

covstats.dpth <- rbind(covstats.dpth, depth.process  %>% group_by(reference)%>% summarise(step="post-filtration depth",minimum = min(after), average=mean(after), median = median(after), maximum = max(after)))

covstats.dpth <- rbind(covstats.dpth, depth.process  %>% group_by(reference) %>% summarise(step="depth retention percent",minimum = min(depth_retention), average=mean(depth_retention), median = median(depth_retention), maximum = max(depth_retention)))

# %>% kable(caption="", digits=1 )



tbl_cnt <- tbl_cnt + 1


covstats.dpth %>% group_by(step)  %>%  gt() %>% fmt_number(columns = c(minimum, average, median, maximum), decimals = 1, rows = step != "depth retention percent") %>% fmt_percent(columns=c(minimum,median,average,maximum), rows = step == "depth retention percent",  decimals=1) %>% tab_header(title=paste("Table ",tbl_cnt, ". Depth of Coverage Statistics for Raw and Filtered Alignments", sep = ""), subtitle="bwa, bwaUniq") 




```

```{r echo=FALSE}
fig_cnt <- fig_cnt + 1

ggplot(depth.process %>%  select(-c(depth_retention,measure)) %>%  gather(before:after, key=measure, value=value) %>% mutate(measure=factor(measure, levels=c("before","after"))) ) +geom_line(aes(group=sample, x=measure,y=value))+ geom_point(aes(group=sample, x=measure,y=value, color=sample))  + facet_grid(.~reference) + labs(title="Depth Of Coverage for Raw and Filtered Alignments", x="", y="Reads Per BP, Genome-Wide" ) + theme_bw()


```

Breadth of coverage, ie, the percentage of the genome covered by at least one read:

```{r echo=FALSE, include=FALSE}
breadth.process <- all_alignments %>%  filter(measure=='total_breadth' ) %>% spread(aligner, value) %>%  mutate(breadth_retention = bwaUniq/bwa) %>% rename(bwa = "before", bwaUniq="after")

covstats.brdth <- breadth.process %>% group_by(reference)%>% summarise(step="pre-filtration breadth",minimum = min(before), average=mean(before), median = median(before), maximum = max(before),reference = unique(reference))

covstats.brdth <- rbind(covstats.brdth, breadth.process %>% summarise(step="post-filtration breadth",minimum = min(after), average=mean(after), median = median(after), maximum = max(after), reference = unique(reference)))

covstats.brdth <- rbind(covstats.brdth, breadth.process  %>% summarise(step="breadth retention percent",minimum = min(breadth_retention), average=mean(breadth_retention), median = median(breadth_retention), maximum = max(breadth_retention), reference = unique(reference)))

tbl_cnt <- tbl_cnt + 1

covstats.brdth %>% group_by(reference) %>% gt() %>% fmt_percent(columns = c(minimum, average, median, maximum), decimals = 1) %>% tab_header(title=paste("Table ",tbl_cnt, ". Breadth of Coverage Statistics for Raw and Filtered Alignments", sep = ""), subtitle=" ")
# %>% kable(caption="", digits=1 )

```

```{r echo=FALSE, include=TRUE}

fig_cnt <- fig_cnt + 1


ggplot(breadth.process %>%  select(-c(breadth_retention, measure)) %>%  gather(before:after, key=measure, value=value) %>% mutate(measure=factor(measure, levels=c("before","after")), value=100*value) ) + geom_point(aes(group=sample, x=measure,y=value, color=sample)) + geom_line(aes(group=sample, x=measure,y=value)) + facet_grid(.~reference) + labs(title= paste("Figure ",fig_cnt,". Depth Of Coverage for Raw and Filtered Alignments", sep=""), x="", y="Percentage of Reference Genome Mapped To" )+ theme_bw()


```

```{r echo=FALSE, include=TRUE}


all.vs_droSim1.bwaUniq.depth_by_chrom.df <- read_delim("data/summaries/intermediate/BAMs/all.vs_droSim1.bwaUniq.dpth_by_chrom", delim = "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(all.vs_droSim1.bwaUniq.depth_by_chrom.df) <- c("sample", "chrom", "measure", "value")
all.vs_droSim1.bwaUniq.depth_by_chrom.df %<>% filter(chrom %in% c( "chr2L", "chr2R", "chr3L", "chr3R", "chr4", "chrX")) %>% filter(measure == "average_depth")




fig_cnt <- fig_cnt + 1


all.vs_droSim1.bwaUniq.depth_by_chrom.gg <- ggplot(all.vs_droSim1.bwaUniq.depth_by_chrom.df ) + geom_point(aes(group=sample, x=chrom,y=value, color=sample)) + geom_line(aes(group=sample, x=chrom,y=value, color=sample)) + labs(title= paste("Figure ",fig_cnt,". Depth Of Coverage by Chromosome", sep=""), x="Chromosome", y="Average Depth of Coverage" )+ theme_bw()


all.vs_droSim1.bwaUniq.depth_by_chrom.gg


png(height =  300, width = 500, filename = paste("results/figures/fig",fig_cnt,"_avgCovDpth_byChromosome.png", sep=""))
all.vs_droSim1.bwaUniq.depth_by_chrom.gg

dev.off()


```

## Called Variants

BWAUniq mappings were used to jointly call variants in VCF format via Freebayes [@Garrison2012] using standard filters.

expand on smrtFreeBayes

```{r echo=FALSE, include=FALSE}

#all_samples.bwaUniq.calledVariants.summaryStats <- read_delim("data/summaries/intermediate/VCFs/all.calledVariants.bwaUniq.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
all_samples.bwaUniq.calledVariants.summaryStats <- read_delim("data/summaries/intermediate/VCFs/all.calledVariants.bwaUniq.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(all_samples.bwaUniq.calledVariants.summaryStats) <-c("group", "caller","refGenome","measure","value")

#all_samples.bwaUniq.calledVariants.summaryStats %<>% separate(group, into = c("caller","group"))

reference_genomes_summary.sprud.df <- reference_genomes_summary.df %>%  spread(measure,value) %>%  select(c(-number_contigs)) 

all_samples.bwaUniq.calledVariants.summaryStats.df <- all_samples.bwaUniq.calledVariants.summaryStats

all_samples.calledVariants.snpCount.df <- inner_join(all_samples.bwaUniq.calledVariants.summaryStats.df %>%  filter(measure=="total_snp_count") %>% spread(measure, value) , reference_genomes_summary.sprud.df, by=c("refGenome"="name"))



```

```{r echo=FALSE,warning=FALSE}
tbl_cnt <- tbl_cnt +1
all_samples.calledVariants.snpCount.gt <- all_samples.calledVariants.snpCount.df %>%  mutate(snp_rate=1000*total_snp_count/number_bases ) %>% select(-c(group))  %>% select(c(caller, refGenome, number_bases, total_snp_count, snp_rate))  %>% gt()%>% fmt_number(c(number_bases, total_snp_count), suffixing = T, decimals = 1) %>% fmt_number(snp_rate, decimals = 1) %>%   cols_label(number_bases="Genome Size (bp)" , total_snp_count = "# SNPs", snp_rate = "SNP rate (per kb)", caller = "variant caller") %>% tab_header(title=paste("Table ",tbl_cnt, ". SNP count and per-KB SNP rate across all samples", sep = ""), subtitle="") 

all_samples.calledVariants.snpCount.gt

```

To build this VCF, `r nrow(data_sets.df %>% select(name) %>% unique())` samples called jointly. However, not all sites were called in all samples (eg, due to coverage differences). The sites had the following group-wide call rate:

```{r echo=FALSE, include=FALSE, warning=FALSE}

all_samples_vs_ncbiMau.stdFB.lmiss <- read_delim("data/summaries/intermediate/VCFs/ncbiMau/stdFreeBayes/all.vs_ncbiMau.bwaUniq.summary.lmiss", "\t", escape_double = FALSE, trim_ws = TRUE)
all_samples_vs_ncbiMau.stdFB.lmiss$refgenome <- "ncbiMau"
all_samples_vs_ncbiMau.stdFB.lmiss$caller <- "stdFreeBayes"

all_samples_vs_droSim1.stdFB.lmiss <- read_delim("data/summaries/intermediate/VCFs/droSim1/stdFreeBayes/all.vs_droSim1.bwaUniq.summary.lmiss", "\t", escape_double = FALSE, trim_ws = TRUE)
all_samples_vs_droSim1.stdFB.lmiss$refgenome <- "droSim1"
all_samples_vs_droSim1.stdFB.lmiss$caller <- "stdFreeBayes"


nsamps <- nrow(data_sets.df %>% filter(subgroups == "all"))


all_samples.lmiss <- rbind(all_samples_vs_ncbiMau.stdFB.lmiss, all_samples_vs_droSim1.stdFB.lmiss) %>% select(c(refgenome, caller, N_MISS)) %>%  mutate(refgenome=as.factor(refgenome), N_PRES=nsamps-N_MISS)

```

```{r echo=FALSE}
fig_cnt <- fig_cnt +1
ggplot(all_samples.lmiss) + geom_freqpoly(aes(x=N_PRES, group=refgenome, color=refgenome), bins=nsamps) + scale_x_continuous(name ="Number Samples",limits=c(1,nsamps), breaks =seq(1,nsamps,1)) + theme_clear() + labs(title=paste("Figure ", fig_cnt, ". Histogram of SNPs by Number of Samples Called At Site", sep=""), y="Number of Sites") #+scale_y_log10()
```

The fraction of jointly called SNPs which are individually callable:

```{r echo=FALSE, include=FALSE}
all_samples_vs_ncbiMau.stdFB.imiss <- read_delim("data/summaries/intermediate/VCFs/ncbiMau/stdFreeBayes/all.vs_ncbiMau.bwaUniq.summary.imiss", "\t", escape_double = FALSE, trim_ws = TRUE)
all_samples_vs_ncbiMau.stdFB.imiss$refgenome <- "ncbiMau"
all_samples_vs_ncbiMau.stdFB.imiss$caller <- "stdFreeBayes"

all_samples_vs_droSim1.stdFB.imiss <- read_delim("data/summaries/intermediate/VCFs/droSim1/stdFreeBayes/all.vs_droSim1.bwaUniq.summary.imiss", "\t", escape_double = FALSE, trim_ws = TRUE)
all_samples_vs_droSim1.stdFB.imiss$refgenome <- "droSim1"
all_samples_vs_droSim1.stdFB.imiss$caller <- "stdFreeBayes"



all_samples.imiss  <- rbind(all_samples_vs_ncbiMau.stdFB.imiss, all_samples_vs_droSim1.stdFB.imiss)  %>%  mutate(name=as.factor(INDV), refgenome=as.factor(refgenome), N_PRES=N_DATA-N_MISS) %>% select(c( name, N_MISS, N_PRES, F_MISS,refgenome))


```

```{r echo=FALSE}

all_samples.imiss.augmented <- inner_join(all_samples.imiss, breadth.process %>%  select(c(sample,reference,after)) %>% rename(after="breadth")%>% mutate(breadth = 100*breadth), by=c("name"="sample", "refgenome"="reference"))


all_samples.imiss.augmented <-inner_join(all_samples.imiss.augmented, depth.process %>%  select(c(sample,reference,after)) %>% rename(after="depth"), by=c("name"="sample", "refgenome"="reference"))

all_samples.imiss.augmented <- all_samples.imiss.augmented %>%  gather(breadth:depth, key="measure", value="value")

```

```{r echo=FALSE}
fig_cnt <-fig_cnt+1
ggplot(all_samples.imiss.augmented) + geom_point(aes(x= value, y=1-F_MISS, color=refgenome)) + facet_grid(.~measure, scales="free_x") + theme_clear() + labs(x="", y="Fraction of SNPs Callable", title=paste("Figure ",fig_cnt,". Jointly Called SNPs Callable per Sample", sep =""), subtitle = "by Breadth and Depth of Coverage")

```

## PsiSeq Algorithm Details

### PsiSeq Classic & PsiSeq2

The first two versions of PsiSeq were similar: reads are aligned to a reference genome and the alignments are compared directly using the pileups. Sites were identified which differed in the parent alignments; ancestry was inferred at each such site in the offspring, according to which parents' allele was present, and assigned a 1 for one parent and a zero for the other. These were then averaged by window: a window with a high average score will be enriched in ancestry from parent 1.

The PsiSeq1 workflow published in @Earley2011, PsiSeq, uses a perl script to directly compare read alignments (as mpileup files) between "control" and selected treatments. The scripts included have minor modifications over the original, such as handling edge cases without error.

The original algorithm from @Earley2011 used mpileups as input; in this way it compares the alignment of the backcrossed hybrid and of one parent species, against the other parent species' reference genome.

Each line of an mpileup is individually examined. For sites meeting basic criteria (eg, the reference base is defined), the aligned bases at the corresponding site are tallied and the base with the most supporting reads is considered the genotype at that site:

```{perl echo=TRUE, include=FALSE, eval=F}

sub snp {
	$match = 0;
#	$read_base =~ s/[\^\$\$fn0-9\!\*]//gi;
	if ($read_base =~ /(\.|\,)/) {
		$match = $read_base =~ s/(\.|\,)//gi;
	}	
#	$match = $1;
	$read_base =~ s/[^agtc]//gi;	
	$mismatch_cvg = length($read_base);
	$counter = 0;
	$a = 0; $t = 0; $c = 0; $g = 0;
	if ( length($read_base) > 1) {
		until ($counter == length($read_base)) {
			$vote = substr($read_base, $counter, 1);
			if ( ($vote eq "A") || ($vote eq "a") ) {
				$a++;
			}
			if ( ($vote eq "T") || ($vote eq "t") ) {
				$t++;
			}
			if ( ($vote eq "C") || ($vote eq "c") ) {
				$c++;
			}
			if ( ($vote eq "G") || ($vote eq "g") ) {
				$g++;
			}

			$counter++;
		}
		#find winner
		$mismatch = "null";
		if ( ($a > $c) && ($a > $t) && ($a > $g) && ($a > $match) ) {
			$Aratio = $a / ($mismatch_cvg + $match);
			$mismatch = "A";
		}
		if ( ($t > $a) && ($t > $c) && ($t > $g) && ($t > $match) ) {
			$Tratio = $t / ($mismatch_cvg + $match);
			$mismatch = "T";
		}
		if ( ($c > $a) && ($c > $t) && ($c > $g) && ($c > $match) ) {
			$Cratio = $c / ($mismatch_cvg + $match);
			$mismatch = "C";
		}
		if ( ($g > $a) && ($g > $t) && ($g > $c) && ($g > $match) ) {
			$Gratio = $g / ($mismatch_cvg + $match);
			$mismatch = "G";
		}	
	} else {
		$mismatch = "null";
	}
}
```

The hybrid alignment is first examined, and a dictionary built containing these site calls. Next, the parent alignment is examined. The genotype at each site is called; if the hybrid has a genotype recorded at this site in the dictionary, they are compared and the site is scored 1 if they are the same and zero otherwise. Notably, if the hybrid has no such site recorded, the site is scored zero.

```{perl echo=TRUE, include=FALSE, eval=F}

	if (exists ($hashish{$contig}{$position}) ) {
		
		@hyb = split('\t', $hashish{$contig}{$position});
		$hyb_ref_base = $hyb[0];
		$hyb_mismatch = $hyb[3];

		if ( ($p_ref_base eq $hyb_ref_base) && ($p_mismatch eq $hyb_mismatch) ) {
			print OUT "$contig\t$position\t1\n";
		} else {
			print OUT "$contig\t$position\t0\n";					
		}
	} else {
			print OUT "$contig\t$position\t0\n";

	}

```

PsiSeq2 was written as an update the the earlier software. This included reorganizing it into a Snakemake (@Rahmann2017,) workflow and shifting the emphasis away from simulated reads. It also aimed to give more nuanced and finely controlled comparison of the alignments. However, its development was ultimately overtaken by PopPsiSeq.

### PopPsiSeq and Allele Frequency Shift

The alignment comparison scripts of PsiSeq 1 and 2 are essentially variant callers which identify sites which mismatch the reference genome to a large degree. This was reasonable for its time but has been rendered obsolete by the development of more sophisticated variant calling algorithms; PopPsiSeq was built around FreeBayes (@Garrison2012). This allows the comparison of groups (eg, replicates of a treatment) whereas earlier comparisons were on an individual basis.

Earlier comparisons were also based on the presence or absence of a fixed variant. By working on a population level, PopPsiSeq is able to use difference in allele frequency between groups (of which fixation is an extreme case). This will hopefully increase statistical power and allow examination of eg polygenic traits.

Once the SNPs were called, the VCF file was split into subsets. The simGFP and mauGFP variants which are treated as ancestral populations. Other subsets represent experimental treatments, such as all simulans backcrosses, all needle-eye mutants, or the one needle-eye simulans backcross.

For each SNP still meeting minimum requirments (biallelic, at most one missing sample)*\<- clean this up*, the subgroup-wide allele frequency was calculated. Using the simGFP and mauGFP as ancestral, the distance to the simulans frequency and the mauritiana frequency calculated for each SNP, for each subset. The per-window average shift was then calculated.

Here is a hypothetical example: suppose that at a given site in the genome, 75% of alleles in the mauritiana population are T and 25% are A. Suppose in the simulans population, it's 25% T and 75% A. Now, the allele frequency is tallied in three different subgroups:

In the first subgroup, 100% of alleles are T. This subgroup would have a mau-ward shift of +0.25 and a sim-ward shift of -0.75.

In the second subgroup, 50% of alleles are T. This subgroup would have a mau-ward shift of -0.25 and a sim-ward shift of -0.25.

In the third subgroup, 0% of alleles are T and all are A. This subgroup would have a mau-ward shift of -0.75 and a sim-ward shift of +0.25.

```{r}
library("ggarchery")
library("patchwork")



illustration.df <- data.frame(example = c(rep("exaggerated mauritiana", 3),rep("intermediate", 3),rep("exaggerated simulans", 3)), population =rep(c("simulans", "mauritiana", "hybrid"),3), allele_frequency = c(0.25, 0.75, 1, 0.25, 0.75, 0.5, 0.25, 0.75, 0))

illustration.df %<>% spread(key = population, value = allele_frequency) %>% mutate(simward_shift = simulans-hybrid, mauward_shift = hybrid - mauritiana) %>% gather(key = "direction", value = "AF_shift", c(simward_shift, mauward_shift)) %>% mutate(example = factor(example, levels = c("exaggerated mauritiana","intermediate","exaggerated simulans")))






illustration.gg <- ggplot(illustration.df)  +scale_color_manual(values=c("black","red","blue"))  + facet_wrap( ~ example ) + coord_cartesian(xlim = c(0.8,1.2)) + theme_clear()  

illustration.gg <- illustration.gg + geom_hline( data = . %>% select(c(example, hybrid)) %>%  unique(), aes( yintercept= hybrid ), linetype = "dashed", color = "black")

illustration.gg <- illustration.gg + geom_point( data = . %>% select(c(example, simulans, mauritiana, hybrid)) %>%  unique()  %>% gather(key = parent, value = AF, c("simulans", "mauritiana", "hybrid")), aes(y= AF , color = parent), x=1,  size = 3) 

illustration.gg <- illustration.gg + geom_hline(aes(yintercept = 0.75), color = "red", linetype = "dashed")+ geom_hline(aes(yintercept = 0.25), color = "blue", linetype = "dashed") 

illustration.gg <- illustration.gg +  geom_arrowsegment(data = . %>% filter(direction == "simward_shift") %>% group_by(example)%>% mutate(arrow_start = max(hybrid, simulans), arrow_end = min(hybrid, simulans)), aes(y = arrow_start, yend =arrow_end) , x = 0.9, xend = 0.9 , color = "blue", arrow_positions =  c(1))


illustration.gg <- illustration.gg +  geom_arrowsegment(data = . %>% filter(direction == "mauward_shift") %>% group_by(example)%>% mutate(arrow_start = min(hybrid, mauritiana), arrow_end = max(hybrid, mauritiana)), aes(y = arrow_start, yend =arrow_end) , x = 1.1, xend = 1.1 , color = "red", arrow_positions =  c(1))


illustration.gg <- illustration.gg + geom_text( data = data.frame(text = c("simulans AF", "mauritiana AF"), x=c(1.1, 0.9), y=c(0.2,0.8), example = c("exaggerated mauritiana","exaggerated simulans") , parent = c("simulans", "mauritiana")) %>% mutate(example = factor(example)), aes(x = x, y=y, color = parent, label = text) ) 


illustration.gg <- illustration.gg + geom_text( data = illustration.df, aes(label=case_match(direction, "simward_shift" ~ paste("simulans-oriented\nAF shift:",AF_shift), "mauward_shift" ~ paste("mauritiana-oriented\nAF shift:",AF_shift)  ), y =case_match(direction, "simward_shift" ~ hybrid + AF_shift, "mauward_shift" ~  hybrid - AF_shift ) , color = case_match(direction, "simward_shift" ~ "simulans", "mauward_shift" ~ "mauritiana"), x = case_match(direction, "simward_shift" ~ 0.85, "mauward_shift" ~ 1.15  )), angle = 90) 

#  library(ggfittext)
  
  

illustration.gg <- illustration.gg + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(),legend.position="none") + labs(y="")


illustration.loSim.gg <- illustration.gg
 

















illustration.df <- data.frame(example = c(rep("exaggerated simulans", 3),rep("intermediate", 3),rep("exaggerated mauritiana", 3)), population =rep(c("mauritiana", "simulans", "hybrid"),3), allele_frequency = c(0.25, 0.75, 1, 0.25, 0.75, 0.5, 0.25, 0.75, 0))

illustration.df %<>% spread(key = population, value = allele_frequency) %>% mutate(simward_shift = hybrid-simulans, mauward_shift = mauritiana-hybrid ) %>% gather(key = "direction", value = "AF_shift", c(simward_shift, mauward_shift)) %>% mutate(example = factor(example, levels = c("exaggerated mauritiana","intermediate","exaggerated simulans")))


illustration.gg <- ggplot(illustration.df)  +scale_color_manual(values=c("black","red","blue"))  + facet_wrap( ~ example ) + coord_cartesian(xlim = c(0.8,1.2)) + theme_clear() + theme( strip.background = element_blank(), strip.text.x = element_blank())



illustration.gg <- illustration.gg + geom_hline( data = . %>% select(c(example, hybrid)) %>%  unique(), aes( yintercept= hybrid ), linetype = "dashed", color = "black")



illustration.gg <- illustration.gg + geom_point( data = . %>% select(c(example, simulans, mauritiana, hybrid)) %>%  unique()  %>% gather(key = parent, value = AF, c("simulans", "mauritiana", "hybrid")), aes(y= AF , color = parent), x=1,  size = 3) 


illustration.gg <- illustration.gg + geom_hline(aes(yintercept = 0.25), color = "red", linetype = "dashed")+ geom_hline(aes(yintercept = 0.75), color = "blue", linetype = "dashed") 


illustration.gg <- illustration.gg +  geom_arrowsegment(data = . %>% filter(direction == "simward_shift") %>% group_by(example)%>% mutate(arrow_start = min(hybrid, simulans), arrow_end = max(hybrid, simulans)), aes(y = arrow_start, yend =arrow_end) , x = 0.9, xend = 0.9 , color = "blue", arrow_positions =  c(1))



illustration.gg <- illustration.gg +  geom_arrowsegment(data = . %>% filter(direction == "mauward_shift") %>% group_by(example)%>% mutate(arrow_start = max(hybrid, mauritiana), arrow_end = min(hybrid, mauritiana)), aes(y = arrow_start, yend =arrow_end) , x = 1.1, xend = 1.1 , color = "red", arrow_positions =  c(1))




illustration.gg <- illustration.gg + geom_text( data = data.frame(text = c("simulans AF", "mauritiana AF"), x=c( 1.1, 0.9),  y=c(0.8, 0.2), example = c("exaggerated mauritiana","exaggerated simulans") , parent = c("simulans", "mauritiana")) %>% mutate(example = factor(example)), aes(x = x, y=y, color = parent, label = text) ) 



illustration.gg <- illustration.gg + geom_text( data = illustration.df, aes(label=case_match(direction, "simward_shift" ~ paste("simulans-oriented\nAF shift:",AF_shift), "mauward_shift" ~ paste("mauritiana-oriented\nAF shift:",AF_shift)  ), y =case_match(direction, "simward_shift" ~ hybrid - AF_shift, "mauward_shift" ~  hybrid + AF_shift ) , color = case_match(direction, "simward_shift" ~ "simulans", "mauward_shift" ~ "mauritiana"), x = case_match(direction, "simward_shift" ~ 0.85, "mauward_shift" ~ 1.15  )), angle = 90) 



illustration.gg <- illustration.gg + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(),legend.position="none") + labs(y="")

illustration.hiSim.gg <- illustration.gg


wrap_elements(illustration.loSim.gg/illustration.hiSim.gg)   +  labs(tag = "Allele Frequency (AF)") + theme(plot.tag = element_text(size = rel(1), angle = 90),plot.tag.position = "left" )


fig_cnt <- fig_cnt + 1
scalar <- 3
png(height =  500*scalar, width = 800*scalar, res = 250, filename = paste("results/figures/fig",fig_cnt,"algorithmExplanationSchematic.png", sep=""))


wrap_elements(illustration.loSim.gg/illustration.hiSim.gg)   +  labs(tag = "Allele Frequency (AF)") + theme(plot.tag = element_text(size = rel(1), angle = 90),plot.tag.position = "left" )



dev.off()


```

## Gene Lists

A curated list of genes of interest was provided by Amanda Moehring (5 Dec 2023):

```         
Lgr4 (CG34411)
Bap60 (CG4303)
DNAlig4 (CG12176)
Fer3HCH (CG4349)
Nna1 (CG44533)
CG2692
CG10996
lncRNA:CR45622 (CR45622)
gce (CG42739)
mh (CG9203)
CG32820
CG32819
Dhc16F (CG7092)
CG15373, CG17450
tilB (CG14620)
p-cup (CG12993)
pcm (CG3291)
r-cup (CG10998)
Nup153 (CG4453) 
Ulp1 (CG12359)
wupA (CG7178)
Ste (FBgn0003523)
Ada3 (CG7098)
CG15446
CG4318
```

Of these, most were easily converted from melanogaster to simulans, using www.orthodb.org, and their gene identifier in the NCBI annotation retrieved from <https://www.ncbi.nlm.nih.gov/gene/>:

```         
Lgr4 (CG34411)    GD15902   LOC6725833
Bap60 (CG4303)    GD15903   LOC6725831
DNAlig4 (CG12176)       GD27483   LOC27207332
Fer3HCH (CG4349)        GD15908   LOC6725819
Nna1 (CG44533)      GD17141   LOC6725884
CG2692      GD24898   LOC6736171
CG10996     GD15872   LOC6725893
gce (CG42739)       GD17204   LOC6726007
mh (CG9203      27209153    LOC27209153
Dhc16F (CG7092)     GD24683   LOC6739854
CG15373     GD17382   LOC6726322
tilB (CG14620)      GD10326   LOC6733227
p-cup (CG12993)     GD17357   LOC6726287
pcm (CG3291)        GD27168   LOC27207018
r-cup (CG10998)     LOC6727351    LOC6727351
Nup153 (CG4453)     GD24841   LOC6740169 
wupA (CG7178)       GD24482   LOC6740188
Ada3 (CG7098)       GD27178   LOC27207028
CG15446     GD24437   LOC6740264
CG4318      GD15906   LOC6725826
```

A handful of genes were refractory: `lncRNA:CR45622` did not appear to have expression or a known transcript in simulans. Stellate, `Ste`, appears to be a family rather than a single gene (use the family as a gene list of its own?) `Ulp1` seems to be associated with two different genes in simulans; `LOC6726423` looks to be the right one. `CG32820`,`CG32819`,`CG17450` are tandem duplicates which are all listed as orthologs of the simulans `LOC6726613`; only one is included in the final list. Finally, gooseberry-neuro, `CG2692`, was found to be on chr2R in melanogaster; so was its ortholog in simulans. The list is meant to be genes on chrX with expression in male reproductive tissue, so this gene was excluded.

```         
LOC6725833  GD15902 Lgr4(CG34411)
LOC6725831  GD15903 Bap60(CG4303)
LOC27207332 GD27483 DNAlig4(CG12176)
LOC6725819  GD15908 Fer3HCH(CG4349)
LOC6725884  GD17141 Nna1(CG44533)
LOC6725893  GD15872 CG10996
LOC6726007  GD17204 gce(CG42739)
LOC27209153 27209153    mh(CG9203
LOC6726613  GD15491 CG32820
LOC6739854  GD24683 Dhc16F(CG7092)
LOC6726322  GD17382 CG15373
LOC6733227  GD10326 tilB(CG14620)
LOC6726287  GD17357 p-cup(CG12993)
LOC27207018 GD27168 pcm(CG3291)
LOC6727351  LOC6727351  r-cup(CG10998)
LOC6740169  GD24841 Nup153(CG4453)
LOC6726423  120285358/GD15595   Ulp1(CG12359)
LOC6740188  GD24482 wupA(CG7178)
LOC27207028 GD27178 Ada3(CG7098)
LOC6740264  GD24437 CG15446
LOC6725826  GD15906 CG4318
```

With this gene list in hand, the corresponding genomic locii were extracted by pulling from the gene annotation GTF these genes and in particular the lines with the "gene" tag in the feature field. These were converted to BED format and extended by 10kb in each direction using the bedtools [@Quinlan2010] slop utility. These locii were used as intervals for averaging the PopPsiSeq frequency the way windows are for whole-genome scans. To sample the background genomic distribution, a gene list's locii were shuffled across their containing contigs without overlap to generate pseudolocii intervals; ie,

```{bash echo=TRUE, eval=FALSE}
bedtools shuffle -chrom -noOverlapping -i {input.gene_bed} -g {fai} > {output.shuff_bed}
```

Ten such shuffles were generated per list. The genetic background was generated similarly, by shuffling the gene annotation and picking the first genes with the same chromosome distribution.

# Results

## PsiSeq Classic

```{r echo=FALSE, warning=FALSE}

BCM10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCM10F.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sample <- "BCM10F"
BCM10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
BCM10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCM10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCM10NE.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sample <- "BCM10NE"
BCM10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sperm <- "mutant"
BCM10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg$backcross <- "mauritiana"




BCS10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCS10F.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sample <- "BCS10F"
BCS10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
BCS10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg$backcross <- "simulans"

BCS10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCS10NE.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sample <- "BCS10NE"
BCS10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sperm <- "mutant"
BCS10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg$backcross <- "simulans"

mauGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/mauGFP.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
mauGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sample <- "mauGFP"
mauGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
mauGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg$backcross <- "none"

simGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/simGFP.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
simGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sample <- "simGFP"
simGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
simGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg$backcross <- "none"








BCM10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCM10F.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCM10F"
BCM10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
BCM10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCM10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCM10NE.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCM10NE"
BCM10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sperm <- "mutant"
BCM10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$backcross <- "mauritiana"




BCS10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCS10F.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCS10F"
BCS10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
BCS10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$backcross <- "simulans"

BCS10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/BCS10NE.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCS10NE"
BCS10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sperm <- "mutant"
BCS10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$backcross <- "simulans"

mauGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/mauGFP.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
mauGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sample <- "mauGFP"
mauGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
mauGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$backcross <- "none"

simGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/droSim1/bwaUniq/simGFP.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
simGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sample <- "simGFP"
simGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
simGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$backcross <- "none"












# 
# BCM10F.PsiSeq.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/ncbiMau/bwaUniq/BCM10F.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
# BCM10NE.PsiSeq.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/ncbiMau/bwaUniq/BCM10NE.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
# BCS10F.PsiSeq.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/ncbiMau/bwaUniq/BCS10F.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
# BCS10NE.PsiSeq.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/ncbiMau/bwaUniq/BCS10NE.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
# mauGFP.PsiSeq.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/ncbiMau/bwaUniq/mauGFP.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
# simGFP.PsiSeq.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq/ncbiMau/bwaUniq/simGFP.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")



all.PsiSeq.shared_with_simGFP.vs_droSim1.bg <- c(BCM10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg, BCM10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg, BCS10F.PsiSeq.shared_with_simGFP.vs_droSim1.bg, BCS10NE.PsiSeq.shared_with_simGFP.vs_droSim1.bg, mauGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg, simGFP.PsiSeq.shared_with_simGFP.vs_droSim1.bg)


all.PsiSeq.shared_with_mauGFP.vs_droSim1.bg <- c(BCM10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg, BCM10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg, BCS10F.PsiSeq.shared_with_mauGFP.vs_droSim1.bg, BCS10NE.PsiSeq.shared_with_mauGFP.vs_droSim1.bg, mauGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg, simGFP.PsiSeq.shared_with_mauGFP.vs_droSim1.bg)

```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}
fig_cnt <- fig_cnt + 1

all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg <- all.PsiSeq.shared_with_simGFP.vs_droSim1.bg[all.PsiSeq.shared_with_simGFP.vs_droSim1.bg@seqnames %in% c('chr2L','chr2R','chr3L','chr3R', "chrX")]

autoplot(all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg, aes(y=ratio, group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x")+ labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,". All Samples Analyzed via PsiSeq1"), subtitle = "fraction of sites where backcrossed flies share an allele with the simGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

png(height =  800, width = 1300, filename = paste("results/figures/fig",fig_cnt,"_Everything_PsiSeq1.png", sep=""))
autoplot(all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg, aes(y=ratio, group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x")+ labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,". All Samples Analyzed via PsiSeq1"), subtitle = "fraction of sites where backcrossed flies share an allele with the simGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))
dev.off()

```

Above is the PsiSeq1 analysis, in which all samples have been compared to simGFP. The vertical axis is the fraction of the sites where the simGFP sample differs from the reference genome, and the sample under analysis also shares this variant allele. As expected, simGFP has a high similarity with itself across the genome, whereas mauritiana and mauritiana backcrosses have low similarity.

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}

letters <- c("a","b","c","d","e","f")
zomes <- c("chr2L","chr2R","chr3L","chr3R","chrX")

figz <- c()

fig_cnt <- fig_cnt + 1




autoplot(all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg[all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c("chr3L") & all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio,group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt, "Analyzed via PsiSeq1: Simulans backcrosses on chromosome 3L "), subtitle = "fraction of sites where backcrossed flies share an allele with the SimGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))



for (x in 1:5) {
png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt, letters[x],"_simBackCross",zomes[x],"_PsiSeq1.png", sep=""))
  
autoplot(all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg[all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c(zomes[x]) & all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio,group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,letters[x], ". Analyzed via PsiSeq1: Simulans backcrosses on chromosome ", zomes[x], sep=""), subtitle = "fraction of sites where backcrossed flies share an allele with the SimGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

dev.off()
  
}

```

Here's a closer look, restricted to the 3L chromosome and just the simulans backcrosses. Two features stand out: a region \~8-10Mb where the wild-type backcross has decreased similarity to simGFP, and a region \~15-20 Mb where the needle-eye backcross has decreased similarity relative to the wild-type backcross.

```{r echo=FALSE, warning=FALSE, include=TRUE}



fig_cnt <- fig_cnt + 1


#all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.chr2L <- all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal[all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal@seqnames=='chr2L']

#freqCompare.windowed.snpCount.tk <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.chr2L, aes(y=num_snp, color=subgroup), geom='line') + theme_clear() + labs(y='')

sim_bkCrss.PsiSeq.shared_with_simGFP.vs_droSim1.simFrq.tk <- autoplot(all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg[all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c("chr3L") & all.PsiSeq.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio, color=sample), geom='line') + theme_clear() + labs(y='') + guides(color="none")#FALSE)

mau_bkCrss.PsiSeq.shared_with_simGFP.vs_droSim1.mauFrq.tk <- autoplot(all.PsiSeq.shared_with_mauGFP.vs_droSim1.bg[all.PsiSeq.shared_with_mauGFP.vs_droSim1.bg@seqnames %in% c("chr3L") & all.PsiSeq.shared_with_mauGFP.vs_droSim1.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio, color=sample), geom='line') + theme_clear() + labs(y='') + guides(color="none")


#"SNPs per Window" = freqCompare.windowed.snpCount.tk,
PsiSeq1.tks.chr3L <- tracks("Simulans Similarity" = sim_bkCrss.PsiSeq.shared_with_simGFP.vs_droSim1.simFrq.tk, "Mauritiana Similiarity" = mau_bkCrss.PsiSeq.shared_with_simGFP.vs_droSim1.mauFrq.tk, title = "Simulans Backcrosses: Similarity to SimGFP and mauGFP, according to PsiSeq1")

PsiSeq1.tks.chr3L



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_extended_simBackCross3L_PsiSeq1.png", sep=""))
PsiSeq1.tks.chr3L
dev.off()

```

Finally, here are the comparisons with both simGFP and mauGFP. It appears that only the second feature near the end of the chromosome arm is the decrease in simulans similarity accompanied by an increase in mauritiana similarity.

## PsiSeq2

```{r echo=FALSE, warning=FALSE, include=FALSE}

BCM10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCM10F.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCM10NE.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCS10F.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCS10NE.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
mauGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/mauGFP.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
simGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/simGFP.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")




BCM10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCM10F.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCM10NE.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCS10F.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/BCS10NE.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
mauGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/mauGFP.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
simGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/droSim1/bwaUniq/simGFP.SNPs_shared_with.mauGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")









BCM10F.PsiSeq2.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/ncbiMau/bwaUniq/BCM10F.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10NE.PsiSeq2.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/ncbiMau/bwaUniq/BCM10NE.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10F.PsiSeq2.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/ncbiMau/bwaUniq/BCS10F.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10NE.PsiSeq2.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/ncbiMau/bwaUniq/BCS10NE.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
mauGFP.PsiSeq2.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/ncbiMau/bwaUniq/mauGFP.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")
simGFP.PsiSeq2.shared_with_simGFP.vs_ncbiMau.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2/ncbiMau/bwaUniq/simGFP.SNPs_shared_with.simGFP.vs_ncbiMau.bwaUniq.genomeWindowed_w100000_s50000.bed")









BCM10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sample <- "BCM10F"
BCM10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
BCM10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCM10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sample <- "BCM10NE"
BCM10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sperm <- "mutant"
BCM10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCS10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sample <- "BCS10F"
BCS10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
BCS10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$backcross <- "simulans"

BCS10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sample <- "BCS10NE"
BCS10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sperm <- "mutant"
BCS10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$backcross <- "simulans"

mauGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sample <- "mauGFP"
mauGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
mauGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$backcross <- "none"

simGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sample <- "simGFP"
simGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
simGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg$backcross <- "none"




all.PsiSeq2.shared_with_simGFP.vs_droSim1.bg <- c(BCM10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg, BCM10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg, BCS10F.PsiSeq2.shared_with_simGFP.vs_droSim1.bg, BCS10NE.PsiSeq2.shared_with_simGFP.vs_droSim1.bg, mauGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg, simGFP.PsiSeq2.shared_with_simGFP.vs_droSim1.bg)





BCM10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCM10F"
BCM10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
BCM10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCM10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCM10NE"
BCM10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sperm <- "mutant"
BCM10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCS10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCS10F"
BCS10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
BCS10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$backcross <- "simulans"

BCS10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sample <- "BCS10NE"
BCS10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sperm <- "mutant"
BCS10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$backcross <- "simulans"

mauGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sample <- "mauGFP"
mauGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
mauGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$backcross <- "none"

simGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sample <- "simGFP"
simGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sperm <- "normal"
simGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$backcross <- "none"




all.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg <- c(BCM10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg, BCM10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg, BCS10F.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg, BCS10NE.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg, mauGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg, simGFP.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg)


```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}


fig_cnt <- fig_cnt + 1


all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg <- all.PsiSeq2.shared_with_simGFP.vs_droSim1.bg[all.PsiSeq2.shared_with_simGFP.vs_droSim1.bg@seqnames %in% c('chr2L','chr2R','chr3L','chr3R', "chrX")]

autoplot(all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg, aes(y=ratio, group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites Shared with SimGFP", title = paste("Figure ", fig_cnt,". All Samples, Analyzed via PsiSeq2"), subtitle = "fraction of sites where backcrossed flies share an allele with the simGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_everything_PsiSeq2.png", sep=""))



autoplot(all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg, aes(y=ratio, group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites Shared with SimGFP", title = paste("Figure ", fig_cnt,". All Samples, Analyzed via PsiSeq2"), subtitle = "fraction of sites where backcrossed flies share an allele with the simGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))


dev.off()


```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}



all.PsiSeq2.shared_with_mauGFP.vs_droSim1.autosomal.bg <- all.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg[all.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg@seqnames %in% c('chr2L','chr2R','chr3L','chr3R', "chrX")]

autoplot(all.PsiSeq2.shared_with_mauGFP.vs_droSim1.autosomal.bg, aes(y=ratio, group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites Shared with SimGFP", title = paste("Figure ", fig_cnt,"b. All Samples, Analyzed via PsiSeq2"), subtitle = "fraction of sites where backcrossed flies share an allele with the mauGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"b_everything_sharedWithMau_PsiSeq2.png", sep=""))



autoplot(all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg, aes(y=ratio, group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites Shared with SimGFP", title = paste("Figure ", fig_cnt,". All Samples, Analyzed via PsiSeq2"), subtitle = "fraction of sites where backcrossed flies share an allele with the simGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))


dev.off()


```

Above is the PsiSeq2 analysis, in which all samples have been compared to simGFP. The vertical axis is the fraction of the sites where the simGFP sample differs from the reference genome, and the sample under analysis also shares this variant allele. As expected, simGFP has a high similarity with itself across the genome, whereas mauritiana the backcrosses' similarity is closer to that of mauGFP. The simulans backcrosses have two distinct behaviors: a "background" in which their similarity is close to that of simGFP, and a "disturbance" on chr3R and intervals on chr2L and chr2R; these generally correspond to similar regions in the PsiSeq1 analysis. As in PsiSeq1, the wild-type backcross is more similar to droSim1 than the needle-eye backcross. What is notable about this analysis is that the backcrosses are actually less similar to droSim1 than mauGFP in the disturbance regions.

Another feature worth mentioning is the deflection on chr2L, \~13-15Mb. This also appears in PsiSeq1 though it is much less pronounced. This looks very similar to one of the peaks which was originally identified in @Earley2011 as a region of interest re: sechellia and simulans speciation; however on closer inspection its actual origins were murky and although it could well be of interest might not mean what we originally thought.

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}



fig_cnt <- fig_cnt + 1

autoplot(all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg[all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c("chr3L") & all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio,group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt, "Analyzed via PsiSeq2: chromosome 3L closeup"), subtitle = "fraction of sites where backcrossed flies share an allele with the SimGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))



for (x in 1:4) {
png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt, letters[x],"_simBackCross",zomes[x],"_PsiSeq2.png", sep=""))
  print(x)

  autoplot(all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg[all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c(zomes[x])  & all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio,group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,letters[x] ,". Analyzed via PsiSeq2: chromosome ",zomes[x]," closeup", sep = ""), subtitle = "fraction of sites where backcrossed flies share an allele with the SimGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() 
dev.off()
  
}




```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"b_chr3L_sharedWithMau_simBackCross",zomes[x],"_PsiSeq2.png", sep=""))


autoplot(all.PsiSeq2.shared_with_mauGFP.vs_droSim1.autosomal.bg[all.PsiSeq2.shared_with_mauGFP.vs_droSim1.autosomal.bg@seqnames %in% c("chr3L") & all.PsiSeq2.shared_with_mauGFP.vs_droSim1.autosomal.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio,group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt, "b. Analyzed via PsiSeq2: chromosome 3L closeup"), subtitle = "fraction of sites where backcrossed flies share an allele with the MauGFP Control", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt, letters[x],"_simBackCross",zomes[x],"_PsiSeq2.png", sep=""))

```

Here's a closer look, restricted to the 3L chromosome and just the simulans backcrosses. As in the PsiSeq1 analysis, there is a region \~15-20 Mb where the needle-eye backcross has decreased similarity relative to the wildtype backcross; however, both backcrosses experience a drop in similarity below that of the mauGFP control.

There is no dramatic deflection in the \~8-10Mb like in the PsiSeq1; however, there is what looks like a local anomaly in the mauGFP genome, where there is an extended interval of heightened similarity to droSim1. This may be causing an artefact in PsiSeq1.

```{r echo=FALSE,warning=FALSE, include=TRUE}



fig_cnt <- fig_cnt + 1

sim_bkCrss.PsiSeq2.shared_with_simGFP.vs_droSim1.simFrq.tk <- autoplot(all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg[all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c("chr3L") & all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio, color=sample), geom='line') + theme_clear() + labs(y='') + guides(color=FALSE)

mau_bkCrss.PsiSeq2.shared_with_simGFP.vs_droSim1.mauFrq.tk <- autoplot(all.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg[all.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg@seqnames %in% c("chr3L") & all.PsiSeq2.shared_with_mauGFP.vs_droSim1.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=ratio, color=sample), geom='line') + theme_clear() + labs(y='') + guides(color=FALSE)


#"SNPs per Window" = freqCompare.windowed.snpCount.tk,
PsiSeq2.tks.chr3L <- tracks("Simulans Similarity" = sim_bkCrss.PsiSeq2.shared_with_simGFP.vs_droSim1.simFrq.tk, "Mauritiana Similiarity" = mau_bkCrss.PsiSeq2.shared_with_simGFP.vs_droSim1.mauFrq.tk, title = paste("Figure ",fig_cnt,". Simulans Backcrosses: Similarity to SimGFP and mauGFP, according to PsiSeq2", sep = ""))

PsiSeq2.tks.chr3L



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_simulansBackCross_3L_PsiSeq2.png", sep=""))
PsiSeq2.tks.chr3L
dev.off()


```

The low similarity of the backcrosses to both parental strains are explained by the fact that PsiSeq2 requires high agreement between reads (ie, it requires the sites to be fixed in both parents and the offspring) whereas the classic algorithm would sample stochastically. However, if we look at the fraction of sites which are heterozygous in each sample, there is still considerable heterozygosity in the hybrids and to a lesser extent the simulans F0

```{r echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}

het_loader <- function(filename){
  
  #filename <- "data/intermediate/population_genetics/heterozygosity/stdFreeBayes/all.vs_droSim1.bwaUniq.het.w100000_s50000.bg"
  
  windowedHet.bg <- import.bedGraph(filename)#, genome=refgen)

  names(mcols(windowedHet.bg)) <-  c("name","score", "strand", "sample", "obs_hom", "exp_hom", "n_sites", "inbreeding_F")

  mcols(windowedHet.bg)[["obs_hom"]] <- as.numeric(mcols(windowedHet.bg)[["obs_hom"]])
  mcols(windowedHet.bg)[["exp_hom"]] <- as.numeric(mcols(windowedHet.bg)[["exp_hom"]])
  mcols(windowedHet.bg)[["n_sites"]] <- as.numeric(mcols(windowedHet.bg)[["n_sites"]])
  mcols(windowedHet.bg)[["inbreeding_F"]] <- as.numeric(mcols(windowedHet.bg)[["inbreeding_F"]])
  
  mcols(windowedHet.bg)[["het_frac"]] <- 1- mcols(windowedHet.bg)[["obs_hom"]]/mcols(windowedHet.bg)[["n_sites"]]

  return(windowedHet.bg)
}

all.vs_droSim1.bwaUniq.het.bg <- het_loader("data/intermediate/population_genetics/heterozygosity/stdFreeBayes/all.vs_droSim1.bwaUniq.het.w100000_s50000.bg")



```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}



fig_cnt <- fig_cnt + 1

autoplot(all.vs_droSim1.bwaUniq.het.bg[all.vs_droSim1.bwaUniq.het.bg@seqnames %in% c("chr3L") & all.vs_droSim1.bwaUniq.het.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=het_frac,group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt, ". Windowed 'Heterozygosity' by Sample"), subtitle = "fraction of SNPs which are 'heterozygous'", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt, "_sample_pseudohet.png", sep=""))

autoplot(all.vs_droSim1.bwaUniq.het.bg[all.vs_droSim1.bwaUniq.het.bg@seqnames %in% c("chr3L") & all.vs_droSim1.bwaUniq.het.bg$sample %in% c("simGFP","mauGFP", "BCS10F" , "BCS10NE")], aes(y=het_frac,group = sample, color = sample), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt, ". Windowed 'Heterozygosity' by Sample"), subtitle = "fraction of SNPs which are 'heterozygous'", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

dev.off()
  




```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}


BCM10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2_relaxed/droSim1/bwaUniq/BCM10F.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCM10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2_relaxed/droSim1/bwaUniq/BCM10NE.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2_relaxed/droSim1/bwaUniq/BCS10F.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
BCS10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2_relaxed/droSim1/bwaUniq/BCS10NE.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
mauGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2_relaxed/droSim1/bwaUniq/mauGFP.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")
simGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg <- psiSeq_loader("data/ultimate/shared_SNPs/PsiSeq2_relaxed/droSim1/bwaUniq/simGFP.SNPs_shared_with.simGFP.vs_droSim1.bwaUniq.genomeWindowed_w100000_s50000.bed")


BCM10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sample <- "BCM10F"
BCM10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
BCM10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCM10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sample <- "BCM10NE"
BCM10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sperm <- "mutant"
BCM10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$backcross <- "mauritiana"

BCS10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sample <- "BCS10F"
BCS10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
BCS10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$backcross <- "simulans"

BCS10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sample <- "BCS10NE"
BCS10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sperm <- "mutant"
BCS10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$backcross <- "simulans"

mauGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sample <- "mauGFP"
mauGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
mauGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$backcross <- "none"

simGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sample <- "simGFP"
simGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$sperm <- "normal"
simGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg$backcross <- "none"




all.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg <- c(BCM10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg, BCM10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg, BCS10F.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg, BCS10NE.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg, mauGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg, simGFP.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg)



all.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.autosomal.bg <- all.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg[all.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.bg@seqnames %in% c('chr2L','chr2R','chr3L','chr3R', "chrX")]



all.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.autosomal.bg$threshold <- "0.5"
all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg$threshold <- "0.9"

thresholds.shared_with_simGFP.vs_droSim1.autosomal.bg <- c(all.PsiSeq2_relaxed.shared_with_simGFP.vs_droSim1.autosomal.bg, all.PsiSeq2.shared_with_simGFP.vs_droSim1.autosomal.bg)

```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}

autoplot(thresholds.shared_with_simGFP.vs_droSim1.autosomal.bg[thresholds.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c("chr3L") & thresholds.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("BCS10F", "BCS10NE", "mauGFP") ], aes(y=ratio, group = threshold, color = threshold), geom='line') + facet_grid(~sample) + labs(y="Fraction of Sites Shared with SimGFP", title = paste("Figure ", fig_cnt,". Impact of Threshold Parameter on PsiSeq2 Analysis"), subtitle = " 50% vs 90% consensus in SNP calling ", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt, "_PsiSeq2_thresholdParameter.png", sep=""))


autoplot(thresholds.shared_with_simGFP.vs_droSim1.autosomal.bg[thresholds.shared_with_simGFP.vs_droSim1.autosomal.bg@seqnames %in% c("chr3L") & thresholds.shared_with_simGFP.vs_droSim1.autosomal.bg$sample %in% c("BCS10F", "BCS10NE", "mauGFP") ], aes(y=ratio, group = threshold, color = threshold), geom='line') + facet_grid(~sample) + labs(y="Fraction of Sites Shared with SimGFP", title = paste("Figure ", fig_cnt,". Impact of Threshold Parameter on PsiSeq2 Analysis"), subtitle = " 50% vs 90% consensus in SNP calling ", x= "coordinate (droSim1 reference genome)") + theme_clear() #+ theme(axis.text.x = element_text(angle = 90, hjust = 1))


dev.off()

```

## Allele Frequency Shift (PopPsiSeq)

Allele frequencies were calculated for all samples and the difference between backcross AF and sim/mauGFP AF was calculated, then summed and averaged by 100kB window.

```{r echo=FALSE, include =FALSE}

all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "mauritiana"

all.sim_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simulans_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.sim_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "simulans"


all.mauBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauritianaBackcross_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mauBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "mauritianaBackcross"

all.simBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simulansBackcross_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "simulansBackcross"


all.mauBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mauBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "mauBackcrossNormal"

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "simBackcrossNormal"





all.mauBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mauBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "mauBackcrossMutant"

all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "simBackcrossMutant"




all.mutantSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mutantSperm_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mutantSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "mutantSperm"


all.normalSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.normalSperm_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.normalSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$subgroup <- "normalSperm"


#all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg <- c(all.mutantSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.normalSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.simBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mauBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mauBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mauBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.sim_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed)


all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg <- c(all.mauBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.simBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.normalSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mutantSperm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mauBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mauBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.sim_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed)



```

```{r echo=FALSE,warning=FALSE}

draw(contrast_samples.hm, column_title=paste("Subgroup Definitions for PopPsiSeq", sep="")) 
```

```{r echo=FALSE,warning=FALSE}
fig_cnt <- fig_cnt + 1


all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal <- all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg[all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg@seqnames %in% c('chr2L','chr2R','chr3L','chr3R',"chrX")]

autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,". Shift in Allele Frequency", sep = ""), subtitle = "Towards Simulans Allele") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_everything_popPsiSeq.png", sep=""))
autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,". Shift in Allele Frequency", sep = ""), subtitle = "Towards Simulans Allele") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))


dev.off()

```

```{r echo=FALSE,warning=FALSE}



library("gtable")
library("cowplot")

shift_legend <- function(p){
  #written by Z. Lin
  # https://stackoverflow.com/questions/54438495/shift-legend-into-empty-facets-of-a-faceted-plot-in-ggplot2/54438496#54438496
  
  
  # check if p is a valid object
  if(!"gtable" %in% class(p)){
    if("ggplot" %in% class(p)){
      gp <- ggplotGrob(p) # convert to grob
    } else {
      message("This is neither a ggplot object nor a grob generated from ggplotGrob. Returning original plot.")
      return(p)
    }
  } else {
    gp <- p
  }

  # check for unfilled facet panels
  facet.panels <- grep("^panel", gp[["layout"]][["name"]])
  empty.facet.panels <- sapply(facet.panels, function(i) "zeroGrob" %in% class(gp[["grobs"]][[i]]))
  empty.facet.panels <- facet.panels[empty.facet.panels]
  if(length(empty.facet.panels) == 0){
    message("There are no unfilled facet panels to shift legend into. Returning original plot.")
    return(p)
  }

  # establish extent of unfilled facet panels (including any axis cells in between)
  empty.facet.panels <- gp[["layout"]][empty.facet.panels, ]
  empty.facet.panels <- list(min(empty.facet.panels[["t"]]), min(empty.facet.panels[["l"]]),
                             max(empty.facet.panels[["b"]]), max(empty.facet.panels[["r"]]))
  names(empty.facet.panels) <- c("t", "l", "b", "r")

  # extract legend & copy over to location of unfilled facet panels
  guide.grob <- which(gp[["layout"]][["name"]] == "guide-box")
  if(length(guide.grob) == 0){
    message("There is no legend present. Returning original plot.")
    return(p)
  }
  gp <- gtable_add_grob(x = gp,
                        grobs = gp[["grobs"]][[guide.grob]],
                        t = empty.facet.panels[["t"]],
                        l = empty.facet.panels[["l"]],
                        b = empty.facet.panels[["b"]],
                        r = empty.facet.panels[["r"]],
                        name = "new-guide-box")

  # squash the original guide box's row / column (whichever applicable)
  # & empty its cell
  guide.grob <- gp[["layout"]][guide.grob, ]
  if(guide.grob[["l"]] == guide.grob[["r"]]){
    gp <- gtable_squash_cols(gp, cols = guide.grob[["l"]])
  }
  if(guide.grob[["t"]] == guide.grob[["b"]]){
    gp <- gtable_squash_rows(gp, rows = guide.grob[["t"]])
  }
  gp <- gtable_remove_grobs(gp, "guide-box")

  return(gp)
}




all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub <- all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal
  

mcols(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub) %<>% as.data.frame() %>% filter(subgroup %in% c("simulans", "mauritiana", "simBackcrossNormal", "simBackcrossMutant", "mauBackcrossNormal", "mauBackcrossMutant" )) %>% mutate(subgroup = case_match(subgroup, "simulans" ~ "simGFP", "mauritiana" ~"mauGFP", "simBackcrossNormal" ~"BCS10WT", "simBackcrossMutant"~"BCS10NE", "mauBackcrossNormal"~"BCM10WT", "mauBackcrossMutant"~"BCM10NE"))

all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub$subgroup <- factor(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub$subgroup, levels = c( "simGFP","BCS10WT", "BCS10NE", "BCM10WT", "BCM10NE", "mauGFP"))


all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub.grob <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub, aes(y=avg_sim_deltaF, color=subgroup), geom='line', linewidth=0.2) + facet_wrap(~seqnames, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,". Shift in Allele Frequency", sep = ""), subtitle = "Towards Simulans Allele") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + scale_color_manual(values=c("simGFP"="red", "BCS10WT"="#ff5e03", "BCS10NE"="#ad9400","BCM10WT"="green", "BCM10NE"="blue", "mauGFP"="purple"))+ guides(color=guide_legend(title=""))



grid.draw(shift_legend(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub.grob@ggplot))


scalar <- 3
png(height =  500*scalar, width = 800*scalar, res = 300, filename = paste("results/figures/fig",fig_cnt,"_selected_popPsiSeq_pretty.png", sep=""))

grid.draw(shift_legend(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal.4pub.grob@ggplot))

dev.off()

```

Above is the output of the PopPsiSeq analysis for all subgroups. There's a lot going on, so let's take a closer look at the simulans backcrosses on chr3L

```{r echo=FALSE,warning=FALSE}

for (x in 1:5) {
png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt, letters[x],"_simBackCross",zomes[x],"_PopPsiSeq.png", sep=""))
  
  
  
  
  print(x)

  autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal[all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal@seqnames %in% c(zomes[x])  & all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal$subgroup %in% c("simulans", "simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,letters[x] ,". Analyzed via PopPsiSeq: chromosome ",zomes[x]," closeup", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate (droSim1 reference genome)", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
  
dev.off()

}

```

```{r echo=FALSE, eval = T, warning=FALSE}

fig_cnt <- fig_cnt + 1

all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.chr3L <- all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal[all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal@seqnames=='chr3L' & all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal$subgroup %in% c("simulans","mauritiana", "simBackcrossMutant", "simBackcrossNormal")]

#c("simulans","mauritiana", "simBackcrossNormal", "simBackcrossMutant", "mauBackcrossNormal")

freqCompare.windowed.snpCount.tk <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.chr3L[all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.chr3L$subgroup == "simulans"], aes(y=num_snp), color="black", geom='line') + theme_clear() + labs(y='')

freqCompare.windowed.simIntrog.tk <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.chr3L, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + theme_clear() + labs(y='')# + guides(color=FALSE)

freqCompare.windowed.mauDeplet.tk <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.chr3L, aes(y=avg_mau_deltaF, color=subgroup), geom='line') + theme_clear() + labs(y='') + guides(color=FALSE)

tks.popPsiSeq.chr3L <- tracks( "SNPs per Window" = freqCompare.windowed.snpCount.tk, "Shift Towards Simulans" = freqCompare.windowed.simIntrog.tk, "Shift Towards Mauritiana" = freqCompare.windowed.mauDeplet.tk, title = paste("Figure ",fig_cnt, ". Comparison of Treatment Flies to Natural Populations\n(chr3L closeup)", sep = ""))

tks.popPsiSeq.chr3L


png(height =  1300, width = 500, filename = paste("results/figures/fig",fig_cnt,"_simulansBackCross_3L_popPsiSeq.png", sep=""))
tks.popPsiSeq.chr3L
dev.off()






```

Here's the chr3L arm with the simulans backcrosses, simulans, and mauritiana. The top panel shows the number of informative SNPs per window. The middle panel shows the average shift towards the simulans allele frequency. The simGFP control in purple is at a baseline of zero, because it's already at the simulans allele frequency! The change in AF for the mauGFP control (red) is negative, as expected. For the most part it is near a value of -1, indicating that most of the SNPs in these region have one allele fixed in simulans and the other allele fixed in maurtiana. There are segments where the mauGFP is elevated above the -1 boundary; this indicates regions where there are many SNPs which have different allele frequencies between simulans and mauritiana, but aren't fixed in both. It is possible that a sample could be lower or higher than one of the parent strains, if alleles are not fixed in one or both, and alelle frequency becomes more extreme than one or the other (eg, if simulans alleles are specifically selected for they might rise to a higher frequency than in the unselected population) However, this is not observed here.

The bottom panel is interpreted similarly, except the change in allele frequency is towards that of MauGFP.

The simulans backcrosses both show values bounded by the simGFP and mauGFP. For most of the chromosome, both backcrosses are mauritiana-like: there would be close to zero change in allele frequency to make the alleles match mauGFP, and the alles would have to be essentially reversed to match simGFP. The same general region \~15-20Mb stands out as in the other methods. Here, both backcrosses are somewhat more simulans-like. The needle-eye mutant is the more simulans-like of the two, being almost heterozygous on average between simGFP and mauGFP; the wildtype is has about half as much simulans character.

There is again an artefact \~8-10Mb, apparently a reduction of allele fixation in one or both of simGFP and mauGFP. The backcrosses behave identically in this blip.

### Window Parameters

```{r echo=FALSE, include =FALSE, eval=T}

all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w1000000_s500000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed$subgroup <- "mauritiana"
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed$window_size <- 1000000

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w1000000_s500000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed$subgroup <- "simBackcrossNormal"
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed$window_size <- 1000000

all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w1000000_s500000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed$subgroup <- "simBackcrossMutant"
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed$window_size <- 1000000



all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w10000_s5000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed$subgroup <- "mauritiana"
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed$window_size <- 10000

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w10000_s5000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed$subgroup <- "simBackcrossNormal"
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed$window_size <- 10000

all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w10000_s5000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed$subgroup <- "simBackcrossMutant"
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed$window_size <- 10000





all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$window_size <- 100000

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$window_size <- 100000

all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed$window_size <- 100000

                                                               

all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousWindows.frqShift.bg <- c(
  
                                                                              
                                                                              all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed,all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed, all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w1000k_s500k.frqShift.bed,
                                                                              
  
  all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed,all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed, all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bed,
  
  all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed,all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed, all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w10k_s5k.frqShift.bed

    )


```

```{r echo=FALSE,warning=FALSE, eval=T}


fig_cnt <- fig_cnt + 1


all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousWindows.frqShift.bg.autosomal <- all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousWindows.frqShift.bg[all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousWindows.frqShift.bg@seqnames %in% c('chr3L')]

autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousWindows.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=log10(window_size), group = window_size), geom='line') + geom_line(data = . %>% filter(window_size == 1000000), color = "black") + facet_wrap(~subgroup, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,"a. Impact of Smoothing Parameters: Genome Window Size", sep = ""), subtitle = "shift averaged over rolling windows 10kB,100kB,500kB, and 1MB wide ") + theme_dark() + theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_color_gradient2(low = "white", mid = "purple", high = "black", midpoint = log10(10000))


png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"a_PopPsiSeq_smoothingParameters_windowed.png", sep=""))
autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousWindows.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=log10(window_size), group = window_size), geom='line') + geom_line(data = . %>% filter(window_size == 1000000), color = "black") + facet_wrap(~subgroup, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,"a. Impact of Smoothing Parameters: Genome Window Size", sep = ""), subtitle = "shift averaged over rolling windows 10kB,100kB,500kB, and 1MB wide ") + theme_dark() + theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_color_gradient2(low = "white", mid = "purple", high = "black", midpoint = log10(10000))


dev.off()

```

at \~12.5 Snp per kb, the standard window is about 1250 SNPs wide...

```{r echo=FALSE, include =FALSE, eval=T}



all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b120_s60.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed$subgroup <- "mauritiana"
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed$bin_size <- 120

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b1250_s625.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed$subgroup <- "simBackcrossNormal"
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed$bin_size <- 120

all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b120_s60.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed$subgroup <- "simBackcrossMutant"
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed$bin_size <- 120








all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b1250_s625.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed$subgroup <- "mauritiana"
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed$bin_size <- 1250

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b1250_s625.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed$subgroup <- "simBackcrossNormal"
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed$bin_size <- 1250

all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b1250_s625.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed$subgroup <- "simBackcrossMutant"
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed$bin_size <- 1250




all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b12500_s6250.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed$subgroup <- "mauritiana"
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed$bin_size <- 12500

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b12500_s6250.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed$subgroup <- "simBackcrossNormal"
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed$bin_size <- 12500

all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/stdFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.snpBinned_b12500_s6250.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed$subgroup <- "simBackcrossMutant"
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed$bin_size <- 12500


all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousSnpBins.frqShift.bg <- c(all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed ,all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed ,all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b120_s60.frqShift.bed ,all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed ,all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed ,all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b1250_s625.frqShift.bed ,all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed ,all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed ,all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.b12500_s6250.frqShift.bed )


```

```{r echo=FALSE,warning=FALSE, eval=T}


#fig_cnt <- fig_cnt + 1


all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousSnpBins.frqShift.bg.autosomal <- all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousSnpBins.frqShift.bg[all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousSnpBins.frqShift.bg@seqnames %in% c('chr3L')]

autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousSnpBins.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=bin_size, group = bin_size), geom='line') + facet_wrap(~subgroup, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,"b. Impact of Smoothing Parameters: SNP Bin Size", sep = ""), subtitle = "shift averaged over rolling bins 120,1250,12500 SNPs wide ") + theme_dark() + theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_color_gradient2(low = "white", mid = "green", high = "black", midpoint = 1250 )


png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"b_PopPsiSeq_smoothingParameters_binned.png", sep=""))


autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.variousSnpBins.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=bin_size, group = bin_size), geom='line') + facet_wrap(~subgroup, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,"b. Impact of Smoothing Parameters: SNP Bin Size", sep = ""), subtitle = "shift averaged over rolling bins 120,1250,12500 SNPs wide ") + theme_dark() + theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_color_gradient2(low = "white", mid = "green", high = "black", midpoint = 1250 )




dev.off()

```

### smrtFreebayes

```{r echo=FALSE, include =FALSE, eval=T}

all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauritiana"

all.sim_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simulans_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.sim_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simulans"


all.mauBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauritianaBackcross_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mauBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauritianaBackcross"

all.simBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simulansBackcross_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simulansBackcross"


all.mauBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mauBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauBackcrossNormal"

all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simBackcrossNormal"





all.mauBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mauBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauBackcrossMutant"

all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simBackcrossMutant"




all.mutantSperm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mutantSperm_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.mutantSperm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mutantSperm"


all.normalSperm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.normalSperm_with_simulans_and_mauritiana.vs_droSim1.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "droSim1", p1 = "sim", p2= "mau")
all.normalSperm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "normalSperm"


all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg <- c(all.mutantSperm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.normalSperm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.simBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mauBckcrss_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mauBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.simBckcrssNorm_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mauBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.simBckcrssMut_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.sim_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mau_with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bed)


```

```{r echo=FALSE,warning=FALSE, eval=T}


fig_cnt <- fig_cnt + 1


all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal <- all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg[all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg@seqnames %in% c('chr2L','chr2R','chr3L','chr3R',"chrX")]

autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,". Shift in Allele Frequency (smrtFreeBayes) ", sep = ""), subtitle = "Towards Simulans Allele") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_everything_popPsiSeq_smrtFB.png", sep=""))
autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,". Shift in Allele Frequency", sep = ""), subtitle = "Towards Simulans Allele") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))


dev.off()

```

```{r echo=FALSE,warning=FALSE, eval=F}



autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal[all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal@seqnames %in% c("chr3L")  & all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal$subgroup %in% c("simulans", "simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,letters[x] ,". Analyzed via PopPsiSeq: chromosome ",zomes[x]," closeup", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate (droSim1 reference genome)", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))




for (x in 1:5) {
png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt, letters[x],"_simBackCross",zomes[x],"_PopPsiSeq.png", sep=""))
  
  
  
  
  print(x)

  autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal[all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal@seqnames %in% c(zomes[x])  & all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal$subgroup %in% c("simulans", "simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,letters[x] ,". Analyzed via PopPsiSeq: chromosome ",zomes[x]," closeup", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate (droSim1 reference genome)", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
  
dev.off()

}

```

```{r echo=FALSE,warning=FALSE, eval=T}

all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal$caller <- "stdFreeBayes"
all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal$caller <- "smrtFreeBayes"


all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal.bothFB <-c(all.with_sim_and_mau.vs_droSim1.bwaUniq.stdFB.w100k_s50k.frqShift.bg.autosomal,all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal)

autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal.bothFB[all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal.bothFB@seqnames %in% c("chr3L")  & all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal.bothFB$subgroup %in% c("simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=caller), geom='line') + facet_wrap(~subgroup) + labs(y="AF Shift", title = paste("Figure ", fig_cnt,". PopPsiSeq: Impact of Variant Calling Strategy ", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate (droSim1 reference genome)", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))


png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_chr3L_popPsiSeq_smrtVsStdFBCompare.png", sep=""))

autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal.bothFB[all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal.bothFB@seqnames %in% c("chr3L")  & all.with_sim_and_mau.vs_droSim1.bwaUniq.w100k_s50k.frqShift.bg.autosomal.bothFB$subgroup %in% c("simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=caller), geom='line') + facet_wrap(~subgroup) + labs(y="AF Shift", title = paste("Figure ", fig_cnt,". PopPsiSeq: Impact of Variant Calling Strategy ", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate (droSim1 reference genome)", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))



dev.off()

```

```{r echo=FALSE,warning=FALSE, eval=F}



autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal[all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal@seqnames %in% c("chr3L")  & all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal$subgroup %in% c("simulans", "simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Fraction of Sites", title = paste("Figure ", fig_cnt,". Analyzed via PopPsiSeq: chromosome ", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate (droSim1 reference genome)", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

```{r echo=FALSE, eval = T, warning=FALSE}

fig_cnt <- fig_cnt + 1

all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.chr3L <- all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal[all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal@seqnames=='chr3L' & all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal$subgroup %in% c("simulans","mauritiana", "simBackcrossNormal", "simBackcrossMutant")]

freqCompare.windowed.snpCount.tk <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.chr3L[all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.chr3L$subgroup == "simulans"], aes(y=num_snp),color="black", geom='line') + theme_clear() + labs(y='')

freqCompare.windowed.simIntrog.tk <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.chr3L, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + theme_clear() + labs(y='') 
freqCompare.windowed.mauDeplet.tk <- autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.chr3L, aes(y=avg_mau_deltaF, color=subgroup), geom='line') + theme_clear() + labs(y='') + guides(color=FALSE)

tks.popPsiSeq.chr3L <- tracks( "SNPs per Window" = freqCompare.windowed.snpCount.tk, "Shift Towards Simulans" = freqCompare.windowed.simIntrog.tk, "Shift Towards Mauritiana" = freqCompare.windowed.mauDeplet.tk, title = paste("Figure ",fig_cnt, ". PopPsiSeq - smrtFreebayes\n(chr3L closeup)", sep = ""))

tks.popPsiSeq.chr3L


png(height =  1300, width = 500, filename = paste("results/figures/fig",fig_cnt,"_simulansBackCross_3L_popPsiSeq.png", sep=""))
tks.popPsiSeq.chr3L
dev.off()






```

### Moving to Princeton Reference Genome

```{r echo=FALSE, include =FALSE, eval=T}

all.mau_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauritiana"

all.sim_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simulans_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.sim_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simulans"


all.mauBckcrss_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauritianaBackcross_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mauBckcrss_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauritianaBackcross"

all.simBckcrss_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simulansBackcross_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.simBckcrss_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simulansBackcross"


all.mauBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mauBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauBackcrossNormal"

all.simBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simBackcrossNormal"





all.mauBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mauBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mauBackcrossMutant"

all.simBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "simBackcrossMutant"




all.mutantSperm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.mutantSperm_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mutantSperm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "mutantSperm"


all.normalSperm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed <- freqShift_loader("data/ultimate/freq_shift/smrtFreeBayes/all.normalSperm_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.windowed_w100000_s50000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.normalSperm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed$subgroup <- "normalSperm"


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bg <- c(all.mutantSperm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.normalSperm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.simBckcrss_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mauBckcrss_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mauBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.simBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mauBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.simBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.sim_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed, all.mau_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bed)

all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg$ref_genome <- "droSim1"

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bg$ref_genome <- "prinDsim3"


all.with_sim_and_mau.reference_genome_compare.bwaUniq.smrtFB.w100k_s50k.frqShift.bg <- c(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg, all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.w100k_s50k.frqShift.bg) 


```

```{r echo=FALSE,warning=FALSE, eval=F}


fig_cnt <- fig_cnt + 1


all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal <- all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg[all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg@seqnames %in% c('chr2L','chr2R','chr3L','chr3R',"chrX")]

autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,". Shift in Allele Frequency (smrtFreeBayes) ", sep = ""), subtitle = "Towards Simulans Allele") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_everything_popPsiSeq_smrtFB.png", sep=""))
autoplot(all.with_sim_and_mau.vs_droSim1.bwaUniq.smrtFB.w100k_s50k.frqShift.bg.autosomal, aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="Simulans-direction AF Shift", title = paste("Figure ",fig_cnt,". Shift in Allele Frequency", sep = ""), subtitle = "Towards Simulans Allele") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))


dev.off()

```

```{r echo=FALSE,warning=FALSE, eval=T}
fig_cnt <- fig_cnt + 1 
autoplot(all.with_sim_and_mau.reference_genome_compare.bwaUniq.smrtFB.w100k_s50k.frqShift.bg[all.with_sim_and_mau.reference_genome_compare.bwaUniq.smrtFB.w100k_s50k.frqShift.bg@seqnames %in% c("chr3L","NC_052522.2")  & all.with_sim_and_mau.reference_genome_compare.bwaUniq.smrtFB.w100k_s50k.frqShift.bg$subgroup %in% c("simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_grid(seqnames~.) + labs(y="AF Shift", title = paste("Figure ", fig_cnt,". PopPsiSeq: Comparison of Simulans Reference Genomes (droSim1 vs Pricenton)", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))

png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"_chr3L_popPsiSeq_refGenomeCompare.png", sep=""))

autoplot(all.with_sim_and_mau.reference_genome_compare.bwaUniq.smrtFB.w100k_s50k.frqShift.bg[all.with_sim_and_mau.reference_genome_compare.bwaUniq.smrtFB.w100k_s50k.frqShift.bg@seqnames %in% c("chr3L","NC_052522.2")  & all.with_sim_and_mau.reference_genome_compare.bwaUniq.smrtFB.w100k_s50k.frqShift.bg$subgroup %in% c("simBackcrossNormal", "simBackcrossMutant", "mauritiana")], aes(y=avg_sim_deltaF, color=subgroup), geom='line') + facet_grid(seqnames~.) + labs(y="AF Shift", title = paste("Figure ", fig_cnt,". PopPsiSeq: Comparison of Simulans Reference Genomes (droSim1 vs Pricenton)", sep = ""), subtitle = "Allele Frequency Shift towards SimGFP", x= "coordinate", y = "average allele frequency change") + theme_clear() + theme(axis.text.x = element_text(angle = 90, hjust = 1))


dev.off()

```

## Gene list

```{r echo=FALSE, include =FALSE, eval=T}



narrow_frqShft <- function(filename, refgen = "dm6", p1 = "sim", p2= "sec"){
  
  tmp.bed <- read_delim(filename, delim = "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
  
  enhancement_col <-  paste("sum_",p1,"_deltaF", sep = "")
  depletion_col <- paste("sum_",p2,"_deltaF", sep="")

  names(tmp.bed) <-  c("chrom","start","stop","name", enhancement_col, depletion_col, "num_snp")

  tmp.bed[[paste("avg_",p1,"_deltaF", sep = "")]] <- tmp.bed[[enhancement_col]]/tmp.bed$num_snp 

  tmp.bed[[paste("avg_",p2,"_deltaF", sep = "")]] <- tmp.bed[[depletion_col]]/tmp.bed$num_snp 

  return(tmp.bed)
}


all.mau_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed <- narrow_frqShft("data/ultimate/freq_shift/smrtFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mau_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed$subgroup <- "mauritiana"

all.sim_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed <- narrow_frqShft("data/ultimate/freq_shift/smrtFreeBayes/all.simulans_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.sim_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed$subgroup <- "simulans"


all.simBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed <- narrow_frqShft("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.simBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed$subgroup <- "simBackcrossNormal"





all.simBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed <- narrow_frqShft("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.simBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed$subgroup <- "simBackcrossMutant"


all.mauBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed <- narrow_frqShft("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mauBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed$subgroup <- "mauBackcrossNormal"





all.mauBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed <- narrow_frqShft("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000.frqShift.bed", refgen = "prinDsim3", p1 = "sim", p2= "mau")
all.mauBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed$subgroup <- "mauBackcrossMutant"



#all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bg <- rbind( all.simBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed, all.simBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed, all.sim_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed, all.mau_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed)


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bg <- rbind( all.simBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed, all.simBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed, all.mauBckcrssNorm_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed, all.mauBckcrssMut_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed, all.mau_with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bed)

```

```{r echo=FALSE, include =FALSE, eval=T}

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg <- data_frame()

for (x in 0:9){
  print(x)
  
tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_f",x,".frqShift.bed", sep =""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "mauritiana"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg, tmp.bed)

#tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.simulans_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_f",x,".frqShift.bed", sep=""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
#tmp.bed$subgroup <- "simulans"
#tmp.bed$rep <- x


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg, tmp.bed)


tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_f",x,".frqShift.bed", sep = ""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "simBackcrossNormal"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg, tmp.bed)



tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_f",x,".frqShift.bed", sep =""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "simBackcrossMutant"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg, tmp.bed)



tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_f",x,".frqShift.bed", sep = ""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "mauBackcrossNormal"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg, tmp.bed)



tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_f",x,".frqShift.bed", sep =""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "mauBackcrossMutant"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg, tmp.bed)


  
  }


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.bg <- all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.bg %>% group_by(subgroup, rep) %>% summarise(avg = mean(avg_sim_deltaF, na.rm = T), std = sd(avg_sim_deltaF, na.rm = T)) 


```

```{r echo=FALSE, include =FALSE, eval=T}

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg <- data_frame()

for (x in 0:9){
  print(x)
  
tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.mauritiana_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_m",x,".frqShift.bed", sep =""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "mauritiana"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg, tmp.bed)

#tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.simulans_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_f",x,".frqShift.bed", sep=""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
#tmp.bed$subgroup <- "simulans"
#tmp.bed$rep <- x


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg, tmp.bed)


tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_m",x,".frqShift.bed", sep = ""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "simBackcrossNormal"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg, tmp.bed)



tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.simBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_m",x,".frqShift.bed", sep =""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "simBackcrossMutant"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg, tmp.bed)



tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossNormal_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_m",x,".frqShift.bed", sep = ""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "mauBackcrossNormal"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg, tmp.bed)



tmp.bed <- narrow_frqShft(paste("data/ultimate/freq_shift/smrtFreeBayes/all.mauBackcrossMutant_with_simulans_and_mauritiana.vs_prinDsim3.bwaUniq.maleExpressionOnX_listWindowed_l10000_m",x,".frqShift.bed", sep =""), refgen = "prinDsim3", p1 = "sim", p2= "mau")
tmp.bed$subgroup <- "mauBackcrossMutant"
tmp.bed$rep <- x

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg <- rbind(all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg, tmp.bed)


  
  }


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.bg <- all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.bg %>% group_by(subgroup, rep) %>% summarise(avg = mean(avg_sim_deltaF, na.rm = T), std = sd(avg_sim_deltaF, na.rm = T)) 


```

```{r echo=FALSE,warning=FALSE, eval=T}


fig_cnt <- fig_cnt + 1 


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.gg <- ggplot() + geom_linerange(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.bg, aes(x=subgroup, ymin = avg - 2*std, ymax = avg + 2*std, group = rep), position = position_dodge(width = 0.5), alpha =0.75) + geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.bg, aes(x=subgroup, y = avg, group = rep), position = position_dodge(width = 0.5), alpha =0.75, color = "black", shape=1) + geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bg , aes(y=avg_sim_deltaF, x=subgroup, color = subgroup), position = position_jitter(width = 0.25, height = 0)) + labs(y="AF Shift (towards simGFP)", title = paste("Figure ", fig_cnt,". Allele Frequency Shift for Genes of Interest", sep = ""), subtitle = "curated list of genes on the X with expression in male sex tissue", x= "", caption = "with 10 replicates of shuffled non-overlapping, same-size intervals") + theme_clear() + theme(axis.text.x = element_text(angle = 30, hjust = 1))  + coord_cartesian(ylim = c(-1,0.5))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.gg



png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"a_Xexpressors_geneList_AFshift_shuffledBkgrnd.png", sep=""))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.gg


dev.off()

```

```{r echo=FALSE,warning=FALSE, eval=T}



all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.pretty4pub.gg <- ggplot() + geom_linerange(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.bg %>% mutate(subgroup = case_match(subgroup, "simulans" ~ "simGFP", "mauritiana" ~"mauGFP", "simBackcrossNormal" ~"BCS10WT", "simBackcrossMutant"~"BCS10NE", "mauBackcrossNormal"~"BCM10WT", "mauBackcrossMutant"~"BCM10NE")), aes(x=subgroup, ymin = avg - 2*std, ymax = avg + 2*std, group = rep), position = position_dodge(width = 0.5),color = "black", alpha =0.6) + geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.bg %>% mutate(subgroup = case_match(subgroup, "simulans" ~ "simGFP", "mauritiana" ~"mauGFP", "simBackcrossNormal" ~"BCS10WT", "simBackcrossMutant"~"BCS10NE", "mauBackcrossNormal"~"BCM10WT", "mauBackcrossMutant"~"BCM10NE")), aes(x=subgroup, y = avg, group = rep), position = position_dodge(width = 0.5), alpha =0.75, color = "black", shape=1) + geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bg %>% mutate(subgroup = case_match(subgroup, "simulans" ~ "simGFP", "mauritiana" ~"mauGFP", "simBackcrossNormal" ~"BCS10WT", "simBackcrossMutant"~"BCS10NE", "mauBackcrossNormal"~"BCM10WT", "mauBackcrossMutant"~"BCM10NE")) , aes(y=avg_sim_deltaF, x=subgroup), color = "black", position = position_jitter(width = 0.1, height = 0)) + labs(y="AF Shift (towards simGFP)", title = paste("Figure ", fig_cnt,". Allele Frequency Shift for Genes of Interest", sep = ""), subtitle = "curated list of genes on the X with expression in male sex tissue", x= "", caption = "with 10 replicates of shuffled non-overlapping, same-size intervals") + theme_clear() + theme(axis.text.x = element_text(angle = 30, hjust = 1))  + coord_cartesian(ylim = c(-1,0.5))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.pretty4pub.gg


scalar <- 3
png(height =  500*scalar, width = 800*scalar, res = 300, filename = paste("results/figures/fig",fig_cnt,"a_Xexpressors_geneList_AFshift_shuffledBkgrnd_pretty.png", sep=""))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.shuffled_controls.frqShift.chonkt.pretty4pub.gg


dev.off()


```

```{r echo=FALSE,warning=FALSE, eval=T}


#fig_cnt <- fig_cnt + 1 


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.gg <- ggplot() + geom_linerange(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.bg, aes(x=subgroup, ymin = avg - 2*std, ymax = avg + 2*std, group = rep), position = position_dodge(width = 0.5), alpha =0.75)+ geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.bg, aes(x=subgroup, y = avg , group = rep), position = position_dodge(width = 0.5), alpha =0.75, shape=1, color = "black") + geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bg , aes(y=avg_sim_deltaF, x=subgroup, color = subgroup), position = position_jitter(width = 0.1, height = 0)) + labs(y="AF Shift (towards simGFP)", title = paste("Figure ", fig_cnt,". Allele Frequency Shift for Genes of Interest", sep = ""), subtitle = "curated list of genes on the X with expression in male sex tissue", x= "", caption = "with 10 replicates of resampled genes from annotation") + theme_clear() + theme(axis.text.x = element_text(angle = 30, hjust = 1))  + coord_cartesian(ylim = c(-1,0.5))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.gg

png(height =  500, width = 800, filename = paste("results/figures/fig",fig_cnt,"b_Xexpressors_geneList_AFshift_resampledBkgrnd.png", sep=""))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.gg


dev.off()

```

```{r echo=FALSE,warning=FALSE, eval=T}


#fig_cnt <- fig_cnt + 1 


all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.pretty4pub.gg <- ggplot() + geom_linerange(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.bg %>% mutate(subgroup = case_match(subgroup, "simulans" ~ "simGFP", "mauritiana" ~"mauGFP", "simBackcrossNormal" ~"BCS10WT", "simBackcrossMutant"~"BCS10NE", "mauBackcrossNormal"~"BCM10WT", "mauBackcrossMutant"~"BCM10NE")), aes(x=subgroup, ymin = avg - 2*std, ymax = avg + 2*std, group = rep), position = position_dodge(width = 0.5), color = "black", alpha =0.6) + geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.bg %>% mutate(subgroup = case_match(subgroup, "simulans" ~ "simGFP", "mauritiana" ~"mauGFP", "simBackcrossNormal" ~"BCS10WT", "simBackcrossMutant"~"BCS10NE", "mauBackcrossNormal"~"BCM10WT", "mauBackcrossMutant"~"BCM10NE")), aes(x=subgroup, y = avg, group = rep), position = position_dodge(width = 0.5), color = "black", alpha =0.6, shape = 1) + geom_point(data = all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.frqShift.bg %>% mutate(subgroup = case_match(subgroup, "simulans" ~ "simGFP", "mauritiana" ~"mauGFP", "simBackcrossNormal" ~"BCS10WT", "simBackcrossMutant"~"BCS10NE", "mauBackcrossNormal"~"BCM10WT", "mauBackcrossMutant"~"BCM10NE")), aes(y=avg_sim_deltaF, x=subgroup), position = position_jitter(width = 0.1, height = 0), color = "black") + labs(y="AF Shift (towards simGFP)", title = paste("Figure ", fig_cnt,". Allele Frequency Shift for Genes of Interest", sep = ""), subtitle = "curated list of genes on the X with expression in male sex tissue", x= "", caption = "with 10 replicates of resampled genes from annotation") + theme_clear() + theme(axis.text.x = element_text(angle = 30, hjust = 1))  + coord_cartesian(ylim = c(-1,0.5))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.pretty4pub.gg


scalar <- 3
png(height =  500*scalar, width = 800*scalar, res = 300, filename = paste("results/figures/fig",fig_cnt,"b_Xexpressors_geneList_AFshift_resampledBkgrnd_pretty.png", sep=""))

all.with_sim_and_mau.vs_prinDsim3.bwaUniq.smrtFB.maleExpressionOnX_l10k.resampled_controls.frqShift.chonkt.pretty4pub.gg

dev.off()

```

# Discussion

# References

## Software

```{r echo=FALSE}
citation("tidyverse")
citation("knitr")
citation("yaml")
```

O. Tange (2018): GNU Parallel 2018, Mar 2018, ISBN 9781387509881, DOI <https://doi.org/10.5281/zenodo.1146014>

## Bibliography
